# Proyecto

En este notebook se explorarán los datos del dataset de **Retail Data Analytics** propuesta en [Kaggle](https://www.kaggle.com/manjeetsingh/retaildataset?select=Features+data+set.csv) mediante el uso de funciones de Spark.

In [0]:
from pyspark.sql.types import StructType, StructField, IntegerType, StringType, ArrayType, FloatType, BooleanType, DoubleType, DateType
from pyspark.sql.functions import to_date, col, datediff, lit
import pyspark.sql.functions as sf

## Leyendo csv's:

### Explorando file system

In [0]:
#display(dbutils.fs.ls("mnt/etlp1a-vicmacbec1@hotmail.com-si/retail-org"))
display(dbutils.fs.ls("FileStore/tables"))

path name size dbfs:/FileStore/tables/Features_data_set-1.csv Features_data_set-1.csv 600478 dbfs:/FileStore/tables/Features_data_set.csv Features_data_set.csv 600478 dbfs:/FileStore/tables/sales_data_set-1.csv sales_data_set-1.csv 13264115 dbfs:/FileStore/tables/sales_data_set.csv sales_data_set.csv 13264115 dbfs:/FileStore/tables/stores_data_set-1.csv stores_data_set-1.csv 577 dbfs:/FileStore/tables/stores_data_set.csv stores_data_set.csv 577

### Features data set.csv

In [0]:
%fs head FileStore/tables/Features_data_set.csv

[Truncated to first 65536 bytes]
Store,Date,Temperature,Fuel_Price,MarkDown1,MarkDown2,MarkDown3,MarkDown4,MarkDown5,CPI,Unemployment,IsHoliday
1,05/02/2010,42.31,2.572,NA,NA,NA,NA,NA,211.0963582,8.106,FALSE
1,12/02/2010,38.51,2.548,NA,NA,NA,NA,NA,211.2421698,8.106,TRUE
1,19/02/2010,39.93,2.514,NA,NA,NA,NA,NA,211.2891429,8.106,FALSE
1,26/02/2010,46.63,2.561,NA,NA,NA,NA,NA,211.3196429,8.106,FALSE
1,05/03/2010,46.5,2.625,NA,NA,NA,NA,NA,211.3501429,8.106,FALSE
1,12/03/2010,57.79,2.667,NA,NA,NA,NA,NA,211.3806429,8.106,FALSE
1,19/03/2010,54.58,2.72,NA,NA,NA,NA,NA,211.215635,8.106,FALSE
1,26/03/2010,51.45,2.732,NA,NA,NA,NA,NA,211.0180424,8.106,FALSE
1,02/04/2010,62.27,2.719,NA,NA,NA,NA,NA,210.8204499,7.808,FALSE
1,09/04/2010,65.86,2.77,NA,NA,NA,NA,NA,210.6228574,7.808,FALSE
1,16/04/2010,66.32,2.808,NA,NA,NA,NA,NA,210.4887,7.808,FALSE
1,23/04/2010,64.84,2.795,NA,NA,NA,NA,NA,210.4391228,7.808,FALSE
1,30/04/2010,67.41,2.78,NA,NA,NA,NA,NA,210.3895456,7.808,FALSE
1,07/05/2010,72.55,2.835,NA,NA,NA,NA,NA,210.3399684,7.808,FALSE
1,14/05/2010,74.78,2.854,NA,NA,NA,NA,NA,210.3374261,7.808,FALSE
1,21/05/2010,76.44,2.826,NA,NA,NA,NA,NA,210.6170934,7.808,FALSE
1,28/05/2010,80.44,2.759,NA,NA,NA,NA,NA,210.8967606,7.808,FALSE
1,04/06/2010,80.69,2.705,NA,NA,NA,NA,NA,211.1764278,7.808,FALSE
1,11/06/2010,80.43,2.668,NA,NA,NA,NA,NA,211.4560951,7.808,FALSE
1,18/06/2010,84.11,2.637,NA,NA,NA,NA,NA,211.4537719,7.808,FALSE
1,25/06/2010,84.34,2.653,NA,NA,NA,NA,NA,211.3386526,7.808,FALSE
1,02/07/2010,80.91,2.669,NA,NA,NA,NA,NA,211.2235333,7.787,FALSE
1,09/07/2010,80.48,2.642,NA,NA,NA,NA,NA,211.108414,7.787,FALSE
1,16/07/2010,83.15,2.623,NA,NA,NA,NA,NA,211.1003854,7.787,FALSE
1,23/07/2010,83.36,2.608,NA,NA,NA,NA,NA,211.2351443,7.787,FALSE
1,30/07/2010,81.84,2.64,NA,NA,NA,NA,NA,211.3699032,7.787,FALSE
1,06/08/2010,87.16,2.627,NA,NA,NA,NA,NA,211.5046621,7.787,FALSE
1,13/08/2010,87,2.692,NA,NA,NA,NA,NA,211.6394211,7.787,FALSE
1,20/08/2010,86.65,2.664,NA,NA,NA,NA,NA,211.6033633,7.787,FALSE
1,27/08/2010,85.22,2.619,NA,NA,NA,NA,NA,211.5673056,7.787,FALSE
1,03/09/2010,81.21,2.577,NA,NA,NA,NA,NA,211.5312479,7.787,FALSE
1,10/09/2010,78.69,2.565,NA,NA,NA,NA,NA,211.4951902,7.787,TRUE
1,17/09/2010,82.11,2.582,NA,NA,NA,NA,NA,211.5224596,7.787,FALSE
1,24/09/2010,80.94,2.624,NA,NA,NA,NA,NA,211.5972246,7.787,FALSE
1,01/10/2010,71.89,2.603,NA,NA,NA,NA,NA,211.6719895,7.838,FALSE
1,08/10/2010,63.93,2.633,NA,NA,NA,NA,NA,211.7467544,7.838,FALSE
1,15/10/2010,67.18,2.72,NA,NA,NA,NA,NA,211.8137436,7.838,FALSE
1,22/10/2010,69.86,2.725,NA,NA,NA,NA,NA,211.8612937,7.838,FALSE
1,29/10/2010,69.64,2.716,NA,NA,NA,NA,NA,211.9088438,7.838,FALSE
1,05/11/2010,58.74,2.689,NA,NA,NA,NA,NA,211.9563939,7.838,FALSE
1,12/11/2010,59.61,2.728,NA,NA,NA,NA,NA,212.003944,7.838,FALSE
1,19/11/2010,51.41,2.771,NA,NA,NA,NA,NA,211.8896737,7.838,FALSE
1,26/11/2010,64.52,2.735,NA,NA,NA,NA,NA,211.7484333,7.838,TRUE
1,03/12/2010,49.27,2.708,NA,NA,NA,NA,NA,211.607193,7.838,FALSE
1,10/12/2010,46.33,2.843,NA,NA,NA,NA,NA,211.4659526,7.838,FALSE
1,17/12/2010,49.84,2.869,NA,NA,NA,NA,NA,211.4053124,7.838,FALSE
1,24/12/2010,52.33,2.886,NA,NA,NA,NA,NA,211.4051222,7.838,FALSE
1,31/12/2010,48.43,2.943,NA,NA,NA,NA,NA,211.4049321,7.838,TRUE
1,07/01/2011,48.27,2.976,NA,NA,NA,NA,NA,211.4047419,7.742,FALSE
1,14/01/2011,35.4,2.983,NA,NA,NA,NA,NA,211.4574109,7.742,FALSE
1,21/01/2011,44.04,3.016,NA,NA,NA,NA,NA,211.8272343,7.742,FALSE
1,28/01/2011,43.83,3.01,NA,NA,NA,NA,NA,212.1970577,7.742,FALSE
1,04/02/2011,42.27,2.989,NA,NA,NA,NA,NA,212.5668812,7.742,FALSE
1,11/02/2011,36.39,3.022,NA,NA,NA,NA,NA,212.9367046,7.742,TRUE
1,18/02/2011,57.36,3.045,NA,NA,NA,NA,NA,213.2478853,7.742,FALSE
1,25/02/2011,62.9,3.065,NA,NA,NA,NA,NA,213.535609,7.742,FALSE
1,04/03/2011,59.58,3.288,NA,NA,NA,NA,NA,213.8233327,7.742,FALSE
1,11/03/2011,53.56,3.459,NA,NA,NA,NA,NA,214.1110564,7.742,FALSE
1,18/03/2011,62.76,3.488,NA,NA,NA,NA,NA,214.3627114,7.742,FALSE
1,25/03/2011,69.97,3.473,NA,NA,NA,NA,NA,214.5

In [0]:
# File location and type
file_location = "/FileStore/tables/Features_data_set.csv"
file_type = "csv"

# CSV options
infer_schema = "true"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
features = (spark.read.format(file_type)
            .option("inferSchema", infer_schema)
            .option("header", first_row_is_header)
            .option("sep", delimiter)
            .load(file_location))

display(features)

Store Date Temperature Fuel_Price MarkDown1 MarkDown2 MarkDown3 MarkDown4 MarkDown5 CPI Unemployment IsHoliday 1 05/02/2010 42.31 2.572 NA NA NA NA NA 211.0963582 8.106 false 1 12/02/2010 38.51 2.548 NA NA NA NA NA 211.2421698 8.106 true 1 19/02/2010 39.93 2.514 NA NA NA NA NA 211.2891429 8.106 false 1 26/02/2010 46.63 2.561 NA NA NA NA NA 211.3196429 8.106 false 1 05/03/2010 46.5 2.625 NA NA NA NA NA 211.3501429 8.106 false 1 12/03/2010 57.79 2.667 NA NA NA NA NA 211.3806429 8.106 false 1 19/03/2010 54.58 2.72 NA NA NA NA NA 211.215635 8.106 false 1 26/03/2010 51.45 2.732 NA NA NA NA NA 211.0180424 8.106 false 1 02/04/2010 62.27 2.719 NA NA NA NA NA 210.8204499 7.808 false 1 09/04/2010 65.86 2.77 NA NA NA NA NA 210.6228574 7.808 false 1 16/04/2010 66.32 2.808 NA NA NA NA NA 210.4887 7.808 false 1 23/04/2010 64.84 2.795 NA NA NA NA NA 210.4391228 7.808 false 1 30/04/2010 67.41 2.78 NA NA NA NA NA 210.3895456 7.808 false 1 07/05/2010 72.55 2.835 NA NA NA NA NA 210.3399684 7.808 false 1 14/05/2010 74.78 2.854 NA NA NA NA NA 210.3374261 7.808 false 1 21/05/2010 76.44 2.826 NA NA NA NA NA 210.6170934 7.808 false 1 28/05/2010 80.44 2.759 NA NA NA NA NA 210.8967606 7.808 false 1 04/06/2010 80.69 2.705 NA NA NA NA NA 211.1764278 7.808 false 1 11/06/2010 80.43 2.668 NA NA NA NA NA 211.4560951 7.808 false 1 18/06/2010 84.11 2.637 NA NA NA NA NA 211.4537719 7.808 false 1 25/06/2010 84.34 2.653 NA NA NA NA NA 211.3386526 7.808 false 1 02/07/2010 80.91 2.669 NA NA NA NA NA 211.2235333 7.787 false 1 09/07/2010 80.48 2.642 NA NA NA NA NA 211.108414 7.787 false 1 16/07/2010 83.15 2.623 NA NA NA NA NA 211.1003854 7.787 false 1 23/07/2010 83.36 2.608 NA NA NA NA NA 211.2351443 7.787 false 1 30/07/2010 81.84 2.64 NA NA NA NA NA 211.3699032 7.787 false 1 06/08/2010 87.16 2.627 NA NA NA NA NA 211.5046621 7.787 false 1 13/08/2010 87.0 2.692 NA NA NA NA NA 211.6394211 7.787 false 1 20/08/2010 86.65 2.664 NA NA NA NA NA 211.6033633 7.787 false 1 27/08/2010 85.22 2.619 NA NA NA NA NA 211.5673056 7.787 false 1 03/09/2010 81.21 2.577 NA NA NA NA NA 211.5312479 7.787 false 1 10/09/2010 78.69 2.565 NA NA NA NA NA 211.4951902 7.787 true 1 17/09/2010 82.11 2.582 NA NA NA NA NA 211.5224596 7.787 false 1 24/09/2010 80.94 2.624 NA NA NA NA NA 211.5972246 7.787 false 1 01/10/2010 71.89 2.603 NA NA NA NA NA 211.6719895 7.838 false 1 08/10/2010 63.93 2.633 NA NA NA NA NA 211.7467544 7.838 false 1 15/10/2010 67.18 2.72 NA NA NA NA NA 211.8137436 7.838 false 1 22/10/2010 69.86 2.725 NA NA NA NA NA 211.8612937 7.838 false 1 29/10/2010 69.64 2.716 NA NA NA NA NA 211.9088438 7.838 false 1 05/11/2010 58.74 2.689 NA NA NA NA NA 211.9563939 7.838 false 1 12/11/2010 59.61 2.728 NA NA NA NA NA 212.003944 7.838 false 1 19/11/2010 51.41 2.771 NA NA NA NA NA 211.8896737 7.838 false 1 26/11/2010 64.52 2.735 NA NA NA NA NA 211.7484333 7.838 true 1 03/12/2010 49.27 2.708 NA NA NA NA NA 211.607193 7.838 false 1 10/12/2010 46.33 2.843 NA NA NA NA NA 211.4659526 7.838 false 1 17/12/2010 49.84 2.869 NA NA NA NA NA 211.4053124 7.838 false 1 24/12/2010 52.33 2.886 NA NA NA NA NA 211.4051222 7.838 false 1 31/12/2010 48.43 2.943 NA NA NA NA NA 211.4049321 7.838 true 1 07/01/2011 48.27 2.976 NA NA NA NA NA 211.4047419 7.742 false 1 14/01/2011 35.4 2.983 NA NA NA NA NA 211.4574109 7.742 false 1 21/01/2011 44.04 3.016 NA NA NA NA NA 211.8272343 7.742 false 1 28/01/2011 43.83 3.01 NA NA NA NA NA 212.1970577 7.742 false 1 04/02/2011 42.27 2.989 NA NA NA NA NA 212.5668812 7.742 false 1 11/02/2011 36.39 3.022 NA NA NA NA NA 212.9367046 7.742 true 1 18/02/2011 57.36 3.045 NA NA NA NA NA 213.2478853 7.742 false 1 25/02/2011 62.9 3.065 NA NA NA NA NA 213.535609 7.742 false 1 04/03/2011 59.58 3.288 NA NA NA NA NA 213.8233327 7.742 false 1 11/03/2011 53.56 3.459 NA NA NA NA NA 214.1110564 7.742 false 1 18/03/2011 62.76 3.488 NA NA NA NA NA 214.3627114 7.742 false 1 25/03/2011 69.97 3.473 NA NA NA NA NA 214.5999389 7.742 false 1 01/04/2011 59.17 3.524 NA NA NA NA NA 214.8371664 7.682 false 1 08/04/

In [0]:
features.printSchema()

root
-- Store: integer (nullable = true)
-- Date: string (nullable = true)
-- Temperature: double (nullable = true)
-- Fuel_Price: double (nullable = true)
-- MarkDown1: string (nullable = true)
-- MarkDown2: string (nullable = true)
-- MarkDown3: string (nullable = true)
-- MarkDown4: string (nullable = true)
-- MarkDown5: string (nullable = true)
-- CPI: string (nullable = true)
-- Unemployment: string (nullable = true)
-- IsHoliday: boolean (nullable = true)

Se observa que Date, MarkDown*, CPI y Unemployment están mal inferidos, por lo que se procede a declarar el esquema.

In [0]:
featuresSchema = StructType([
  StructField("Store", IntegerType(), True), 
  StructField("Date", StringType(), True),
  StructField("Temperature", DoubleType(), True),
  StructField("Fuel_Price", DoubleType(), True),
  StructField("MarkDown1", DoubleType(), True),
  StructField("MarkDown2", DoubleType(), True),
  StructField("MarkDown3", DoubleType(), True),
  StructField("MarkDown4", DoubleType(), True),
  StructField("MarkDown5", DoubleType(), True),
  StructField("CPI", DoubleType(), True),
  StructField("Unemployment", DoubleType(), True),
  StructField("IsHoliday", BooleanType(), True)
])

features2 = (spark.read
             .format(file_type)
             .option("header", first_row_is_header)
             .option("sep", delimiter)
             .schema(featuresSchema)
             .load(file_location))

display(features2)

Store Date Temperature Fuel_Price MarkDown1 MarkDown2 MarkDown3 MarkDown4 MarkDown5 CPI Unemployment IsHoliday 1 05/02/2010 42.31 2.572 null null null null null 211.0963582 8.106 false 1 12/02/2010 38.51 2.548 null null null null null 211.2421698 8.106 true 1 19/02/2010 39.93 2.514 null null null null null 211.2891429 8.106 false 1 26/02/2010 46.63 2.561 null null null null null 211.3196429 8.106 false 1 05/03/2010 46.5 2.625 null null null null null 211.3501429 8.106 false 1 12/03/2010 57.79 2.667 null null null null null 211.3806429 8.106 false 1 19/03/2010 54.58 2.72 null null null null null 211.215635 8.106 false 1 26/03/2010 51.45 2.732 null null null null null 211.0180424 8.106 false 1 02/04/2010 62.27 2.719 null null null null null 210.8204499 7.808 false 1 09/04/2010 65.86 2.77 null null null null null 210.6228574 7.808 false 1 16/04/2010 66.32 2.808 null null null null null 210.4887 7.808 false 1 23/04/2010 64.84 2.795 null null null null null 210.4391228 7.808 false 1 30/04/2010 67.41 2.78 null null null null null 210.3895456 7.808 false 1 07/05/2010 72.55 2.835 null null null null null 210.3399684 7.808 false 1 14/05/2010 74.78 2.854 null null null null null 210.3374261 7.808 false 1 21/05/2010 76.44 2.826 null null null null null 210.6170934 7.808 false 1 28/05/2010 80.44 2.759 null null null null null 210.8967606 7.808 false 1 04/06/2010 80.69 2.705 null null null null null 211.1764278 7.808 false 1 11/06/2010 80.43 2.668 null null null null null 211.4560951 7.808 false 1 18/06/2010 84.11 2.637 null null null null null 211.4537719 7.808 false 1 25/06/2010 84.34 2.653 null null null null null 211.3386526 7.808 false 1 02/07/2010 80.91 2.669 null null null null null 211.2235333 7.787 false 1 09/07/2010 80.48 2.642 null null null null null 211.108414 7.787 false 1 16/07/2010 83.15 2.623 null null null null null 211.1003854 7.787 false 1 23/07/2010 83.36 2.608 null null null null null 211.2351443 7.787 false 1 30/07/2010 81.84 2.64 null null null null null 211.3699032 7.787 false 1 06/08/2010 87.16 2.627 null null null null null 211.5046621 7.787 false 1 13/08/2010 87.0 2.692 null null null null null 211.6394211 7.787 false 1 20/08/2010 86.65 2.664 null null null null null 211.6033633 7.787 false 1 27/08/2010 85.22 2.619 null null null null null 211.5673056 7.787 false 1 03/09/2010 81.21 2.577 null null null null null 211.5312479 7.787 false 1 10/09/2010 78.69 2.565 null null null null null 211.4951902 7.787 true 1 17/09/2010 82.11 2.582 null null null null null 211.5224596 7.787 false 1 24/09/2010 80.94 2.624 null null null null null 211.5972246 7.787 false 1 01/10/2010 71.89 2.603 null null null null null 211.6719895 7.838 false 1 08/10/2010 63.93 2.633 null null null null null 211.7467544 7.838 false 1 15/10/2010 67.18 2.72 null null null null null 211.8137436 7.838 false 1 22/10/2010 69.86 2.725 null null null null null 211.8612937 7.838 false 1 29/10/2010 69.64 2.716 null null null null null 211.9088438 7.838 false 1 05/11/2010 58.74 2.689 null null null null null 211.9563939 7.838 false 1 12/11/2010 59.61 2.728 null null null null null 212.003944 7.838 false 1 19/11/2010 51.41 2.771 null null null null null 211.8896737 7.838 false 1 26/11/2010 64.52 2.735 null null null null null 211.7484333 7.838 true 1 03/12/2010 49.27 2.708 null null null null null 211.607193 7.838 false 1 10/12/2010 46.33 2.843 null null null null null 211.4659526 7.838 false 1 17/12/2010 49.84 2.869 null null null null null 211.4053124 7.838 false 1 24/12/2010 52.33 2.886 null null null null null 211.4051222 7.838 false 1 31/12/2010 48.43 2.943 null null null null null 211.4049321 7.838 true 1 07/01/2011 48.27 2.976 null null null null null 211.4047419 7.742 false 1 14/01/2011 35.4 2.983 null null null null null 211.4574109 7.742 false 1 21/01/2011 44.04 3.016 null null null null null 211.8272343 7.742 false 1 28/01/2011 43.83 3.01 null null null null null 212.1970577 7.742 false 1 04/02/2011 42.27 2.989 null null null null null 212.5668812

Cambiando el tipo de Date (String -> Date)

In [0]:
features3 = features2.withColumn("Date", to_date(col("Date"),"dd/MM/yyyy"))
#changedTypedf = joindf.withColumn("show", joindf["show"].cast(DoubleType()))
display(features3)

Store Date Temperature Fuel_Price MarkDown1 MarkDown2 MarkDown3 MarkDown4 MarkDown5 CPI Unemployment IsHoliday 1 2010-02-05 42.31 2.572 null null null null null 211.0963582 8.106 false 1 2010-02-12 38.51 2.548 null null null null null 211.2421698 8.106 true 1 2010-02-19 39.93 2.514 null null null null null 211.2891429 8.106 false 1 2010-02-26 46.63 2.561 null null null null null 211.3196429 8.106 false 1 2010-03-05 46.5 2.625 null null null null null 211.3501429 8.106 false 1 2010-03-12 57.79 2.667 null null null null null 211.3806429 8.106 false 1 2010-03-19 54.58 2.72 null null null null null 211.215635 8.106 false 1 2010-03-26 51.45 2.732 null null null null null 211.0180424 8.106 false 1 2010-04-02 62.27 2.719 null null null null null 210.8204499 7.808 false 1 2010-04-09 65.86 2.77 null null null null null 210.6228574 7.808 false 1 2010-04-16 66.32 2.808 null null null null null 210.4887 7.808 false 1 2010-04-23 64.84 2.795 null null null null null 210.4391228 7.808 false 1 2010-04-30 67.41 2.78 null null null null null 210.3895456 7.808 false 1 2010-05-07 72.55 2.835 null null null null null 210.3399684 7.808 false 1 2010-05-14 74.78 2.854 null null null null null 210.3374261 7.808 false 1 2010-05-21 76.44 2.826 null null null null null 210.6170934 7.808 false 1 2010-05-28 80.44 2.759 null null null null null 210.8967606 7.808 false 1 2010-06-04 80.69 2.705 null null null null null 211.1764278 7.808 false 1 2010-06-11 80.43 2.668 null null null null null 211.4560951 7.808 false 1 2010-06-18 84.11 2.637 null null null null null 211.4537719 7.808 false 1 2010-06-25 84.34 2.653 null null null null null 211.3386526 7.808 false 1 2010-07-02 80.91 2.669 null null null null null 211.2235333 7.787 false 1 2010-07-09 80.48 2.642 null null null null null 211.108414 7.787 false 1 2010-07-16 83.15 2.623 null null null null null 211.1003854 7.787 false 1 2010-07-23 83.36 2.608 null null null null null 211.2351443 7.787 false 1 2010-07-30 81.84 2.64 null null null null null 211.3699032 7.787 false 1 2010-08-06 87.16 2.627 null null null null null 211.5046621 7.787 false 1 2010-08-13 87.0 2.692 null null null null null 211.6394211 7.787 false 1 2010-08-20 86.65 2.664 null null null null null 211.6033633 7.787 false 1 2010-08-27 85.22 2.619 null null null null null 211.5673056 7.787 false 1 2010-09-03 81.21 2.577 null null null null null 211.5312479 7.787 false 1 2010-09-10 78.69 2.565 null null null null null 211.4951902 7.787 true 1 2010-09-17 82.11 2.582 null null null null null 211.5224596 7.787 false 1 2010-09-24 80.94 2.624 null null null null null 211.5972246 7.787 false 1 2010-10-01 71.89 2.603 null null null null null 211.6719895 7.838 false 1 2010-10-08 63.93 2.633 null null null null null 211.7467544 7.838 false 1 2010-10-15 67.18 2.72 null null null null null 211.8137436 7.838 false 1 2010-10-22 69.86 2.725 null null null null null 211.8612937 7.838 false 1 2010-10-29 69.64 2.716 null null null null null 211.9088438 7.838 false 1 2010-11-05 58.74 2.689 null null null null null 211.9563939 7.838 false 1 2010-11-12 59.61 2.728 null null null null null 212.003944 7.838 false 1 2010-11-19 51.41 2.771 null null null null null 211.8896737 7.838 false 1 2010-11-26 64.52 2.735 null null null null null 211.7484333 7.838 true 1 2010-12-03 49.27 2.708 null null null null null 211.607193 7.838 false 1 2010-12-10 46.33 2.843 null null null null null 211.4659526 7.838 false 1 2010-12-17 49.84 2.869 null null null null null 211.4053124 7.838 false 1 2010-12-24 52.33 2.886 null null null null null 211.4051222 7.838 false 1 2010-12-31 48.43 2.943 null null null null null 211.4049321 7.838 true 1 2011-01-07 48.27 2.976 null null null null null 211.4047419 7.742 false 1 2011-01-14 35.4 2.983 null null null null null 211.4574109 7.742 false 1 2011-01-21 44.04 3.016 null null null null null 211.8272343 7.742 false 1 2011-01-28 43.83 3.01 null null null null null 212.1970577 7.742 false 1 2011-02-04 42.27 2.989 null null null null null 212.5668812

In [0]:
features3.printSchema()

root
-- Store: integer (nullable = true)
-- Date: date (nullable = true)
-- Temperature: double (nullable = true)
-- Fuel_Price: double (nullable = true)
-- MarkDown1: double (nullable = true)
-- MarkDown2: double (nullable = true)
-- MarkDown3: double (nullable = true)
-- MarkDown4: double (nullable = true)
-- MarkDown5: double (nullable = true)
-- CPI: double (nullable = true)
-- Unemployment: double (nullable = true)
-- IsHoliday: boolean (nullable = true)

In [0]:
features3.count()

Out[9]: 8190

### sale data-set.csv

In [0]:
%fs head FileStore/tables/sales_data_set.csv

[Truncated to first 65536 bytes]
Store,Dept,Date,Weekly_Sales,IsHoliday
1,1,05/02/2010,24924.5,FALSE
1,1,12/02/2010,46039.49,TRUE
1,1,19/02/2010,41595.55,FALSE
1,1,26/02/2010,19403.54,FALSE
1,1,05/03/2010,21827.9,FALSE
1,1,12/03/2010,21043.39,FALSE
1,1,19/03/2010,22136.64,FALSE
1,1,26/03/2010,26229.21,FALSE
1,1,02/04/2010,57258.43,FALSE
1,1,09/04/2010,42960.91,FALSE
1,1,16/04/2010,17596.96,FALSE
1,1,23/04/2010,16145.35,FALSE
1,1,30/04/2010,16555.11,FALSE
1,1,07/05/2010,17413.94,FALSE
1,1,14/05/2010,18926.74,FALSE
1,1,21/05/2010,14773.04,FALSE
1,1,28/05/2010,15580.43,FALSE
1,1,04/06/2010,17558.09,FALSE
1,1,11/06/2010,16637.62,FALSE
1,1,18/06/2010,16216.27,FALSE
1,1,25/06/2010,16328.72,FALSE
1,1,02/07/2010,16333.14,FALSE
1,1,09/07/2010,17688.76,FALSE
1,1,16/07/2010,17150.84,FALSE
1,1,23/07/2010,15360.45,FALSE
1,1,30/07/2010,15381.82,FALSE
1,1,06/08/2010,17508.41,FALSE
1,1,13/08/2010,15536.4,FALSE
1,1,20/08/2010,15740.13,FALSE
1,1,27/08/2010,15793.87,FALSE
1,1,03/09/2010,16241.78,FALSE
1,1,10/09/2010,18194.74,TRUE
1,1,17/09/2010,19354.23,FALSE
1,1,24/09/2010,18122.52,FALSE
1,1,01/10/2010,20094.19,FALSE
1,1,08/10/2010,23388.03,FALSE
1,1,15/10/2010,26978.34,FALSE
1,1,22/10/2010,25543.04,FALSE
1,1,29/10/2010,38640.93,FALSE
1,1,05/11/2010,34238.88,FALSE
1,1,12/11/2010,19549.39,FALSE
1,1,19/11/2010,19552.84,FALSE
1,1,26/11/2010,18820.29,TRUE
1,1,03/12/2010,22517.56,FALSE
1,1,10/12/2010,31497.65,FALSE
1,1,17/12/2010,44912.86,FALSE
1,1,24/12/2010,55931.23,FALSE
1,1,31/12/2010,19124.58,TRUE
1,1,07/01/2011,15984.24,FALSE
1,1,14/01/2011,17359.7,FALSE
1,1,21/01/2011,17341.47,FALSE
1,1,28/01/2011,18461.18,FALSE
1,1,04/02/2011,21665.76,FALSE
1,1,11/02/2011,37887.17,TRUE
1,1,18/02/2011,46845.87,FALSE
1,1,25/02/2011,19363.83,FALSE
1,1,04/03/2011,20327.61,FALSE
1,1,11/03/2011,21280.4,FALSE
1,1,18/03/2011,20334.23,FALSE
1,1,25/03/2011,20881.1,FALSE
1,1,01/04/2011,20398.09,FALSE
1,1,08/04/2011,23873.79,FALSE
1,1,15/04/2011,28762.37,FALSE
1,1,22/04/2011,50510.31,FALSE
1,1,29/04/2011,41512.39,FALSE
1,1,06/05/2011,20138.19,FALSE
1,1,13/05/2011,17235.15,FALSE
1,1,20/05/2011,15136.78,FALSE
1,1,27/05/2011,15741.6,FALSE
1,1,03/06/2011,16434.15,FALSE
1,1,10/06/2011,15883.52,FALSE
1,1,17/06/2011,14978.09,FALSE
1,1,24/06/2011,15682.81,FALSE
1,1,01/07/2011,15363.5,FALSE
1,1,08/07/2011,16148.87,FALSE
1,1,15/07/2011,15654.85,FALSE
1,1,22/07/2011,15766.6,FALSE
1,1,29/07/2011,15922.41,FALSE
1,1,05/08/2011,15295.55,FALSE
1,1,12/08/2011,14539.79,FALSE
1,1,19/08/2011,14689.24,FALSE
1,1,26/08/2011,14537.37,FALSE
1,1,02/09/2011,15277.27,FALSE
1,1,09/09/2011,17746.68,TRUE
1,1,16/09/2011,18535.48,FALSE
1,1,23/09/2011,17859.3,FALSE
1,1,30/09/2011,18337.68,FALSE
1,1,07/10/2011,20797.58,FALSE
1,1,14/10/2011,23077.55,FALSE
1,1,21/10/2011,23351.8,FALSE
1,1,28/10/2011,31579.9,FALSE
1,1,04/11/2011,39886.06,FALSE
1,1,11/11/2011,18689.54,FALSE
1,1,18/11/2011,19050.66,FALSE
1,1,25/11/2011,20911.25,TRUE
1,1,02/12/2011,25293.49,FALSE
1,1,09/12/2011,33305.92,FALSE
1,1,16/12/2011,45773.03,FALSE
1,1,23/12/2011,46788.75,FALSE
1,1,30/12/2011,23350.88,TRUE
1,1,06/01/2012,16567.69,FALSE
1,1,13/01/2012,16894.4,FALSE
1,1,20/01/2012,18365.1,FALSE
1,1,27/01/2012,18378.16,FALSE
1,1,03/02/2012,23510.49,FALSE
1,1,10/02/2012,36988.49,TRUE
1,1,17/02/2012,54060.1,FALSE
1,1,24/02/2012,20124.22,FALSE
1,1,02/03/2012,20113.03,FALSE
1,1,09/03/2012,21140.07,FALSE
1,1,16/03/2012,22366.88,FALSE
1,1,23/03/2012,22107.7,FALSE
1,1,30/03/2012,28952.86,FALSE
1,1,06/04/2012,57592.12,FALSE
1,1,13/04/2012,34684.21,FALSE
1,1,20/04/2012,16976.19,FALSE
1,1,27/04/2012,16347.6,FALSE
1,1,04/05/2012,17147.44,FALSE
1,1,11/05/2012,18164.2,FALSE
1,1,18/05/2012,18517.79,FALSE
1,1,25/05/2012,16963.55,FALSE
1,1,01/06/2012,16065.49,FALSE
1,1,08/06/2012,17666,FALSE
1,1,15/06/2012,17558.82,FALSE
1,1,22/06/2012,16633.41,FALSE
1,1,29/06/2012,15722.82,FALSE
1,1,06/07/2012,17823.37,FALSE
1,1,13/07/2012,16566

In [0]:
# File location and type
file_location = "/FileStore/tables/sales_data_set.csv"
file_type = "csv"

# CSV options
infer_schema = "true"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
sales = (spark.read.format(file_type)
         .option("inferSchema", infer_schema)
         .option("header", first_row_is_header)
         .option("sep", delimiter)
         .load(file_location))

display(sales)

Store Dept Date Weekly_Sales IsHoliday 1 1 05/02/2010 24924.5 false 1 1 12/02/2010 46039.49 true 1 1 19/02/2010 41595.55 false 1 1 26/02/2010 19403.54 false 1 1 05/03/2010 21827.9 false 1 1 12/03/2010 21043.39 false 1 1 19/03/2010 22136.64 false 1 1 26/03/2010 26229.21 false 1 1 02/04/2010 57258.43 false 1 1 09/04/2010 42960.91 false 1 1 16/04/2010 17596.96 false 1 1 23/04/2010 16145.35 false 1 1 30/04/2010 16555.11 false 1 1 07/05/2010 17413.94 false 1 1 14/05/2010 18926.74 false 1 1 21/05/2010 14773.04 false 1 1 28/05/2010 15580.43 false 1 1 04/06/2010 17558.09 false 1 1 11/06/2010 16637.62 false 1 1 18/06/2010 16216.27 false 1 1 25/06/2010 16328.72 false 1 1 02/07/2010 16333.14 false 1 1 09/07/2010 17688.76 false 1 1 16/07/2010 17150.84 false 1 1 23/07/2010 15360.45 false 1 1 30/07/2010 15381.82 false 1 1 06/08/2010 17508.41 false 1 1 13/08/2010 15536.4 false 1 1 20/08/2010 15740.13 false 1 1 27/08/2010 15793.87 false 1 1 03/09/2010 16241.78 false 1 1 10/09/2010 18194.74 true 1 1 17/09/2010 19354.23 false 1 1 24/09/2010 18122.52 false 1 1 01/10/2010 20094.19 false 1 1 08/10/2010 23388.03 false 1 1 15/10/2010 26978.34 false 1 1 22/10/2010 25543.04 false 1 1 29/10/2010 38640.93 false 1 1 05/11/2010 34238.88 false 1 1 12/11/2010 19549.39 false 1 1 19/11/2010 19552.84 false 1 1 26/11/2010 18820.29 true 1 1 03/12/2010 22517.56 false 1 1 10/12/2010 31497.65 false 1 1 17/12/2010 44912.86 false 1 1 24/12/2010 55931.23 false 1 1 31/12/2010 19124.58 true 1 1 07/01/2011 15984.24 false 1 1 14/01/2011 17359.7 false 1 1 21/01/2011 17341.47 false 1 1 28/01/2011 18461.18 false 1 1 04/02/2011 21665.76 false 1 1 11/02/2011 37887.17 true 1 1 18/02/2011 46845.87 false 1 1 25/02/2011 19363.83 false 1 1 04/03/2011 20327.61 false 1 1 11/03/2011 21280.4 false 1 1 18/03/2011 20334.23 false 1 1 25/03/2011 20881.1 false 1 1 01/04/2011 20398.09 false 1 1 08/04/2011 23873.79 false 1 1 15/04/2011 28762.37 false 1 1 22/04/2011 50510.31 false 1 1 29/04/2011 41512.39 false 1 1 06/05/2011 20138.19 false 1 1 13/05/2011 17235.15 false 1 1 20/05/2011 15136.78 false 1 1 27/05/2011 15741.6 false 1 1 03/06/2011 16434.15 false 1 1 10/06/2011 15883.52 false 1 1 17/06/2011 14978.09 false 1 1 24/06/2011 15682.81 false 1 1 01/07/2011 15363.5 false 1 1 08/07/2011 16148.87 false 1 1 15/07/2011 15654.85 false 1 1 22/07/2011 15766.6 false 1 1 29/07/2011 15922.41 false 1 1 05/08/2011 15295.55 false 1 1 12/08/2011 14539.79 false 1 1 19/08/2011 14689.24 false 1 1 26/08/2011 14537.37 false 1 1 02/09/2011 15277.27 false 1 1 09/09/2011 17746.68 true 1 1 16/09/2011 18535.48 false 1 1 23/09/2011 17859.3 false 1 1 30/09/2011 18337.68 false 1 1 07/10/2011 20797.58 false 1 1 14/10/2011 23077.55 false 1 1 21/10/2011 23351.8 false 1 1 28/10/2011 31579.9 false 1 1 04/11/2011 39886.06 false 1 1 11/11/2011 18689.54 false 1 1 18/11/2011 19050.66 false 1 1 25/11/2011 20911.25 true 1 1 02/12/2011 25293.49 false 1 1 09/12/2011 33305.92 false 1 1 16/12/2011 45773.03 false 1 1 23/12/2011 46788.75 false 1 1 30/12/2011 23350.88 true 1 1 06/01/2012 16567.69 false 1 1 13/01/2012 16894.4 false 1 1 20/01/2012 18365.1 false 1 1 27/01/2012 18378.16 false 1 1 03/02/2012 23510.49 false 1 1 10/02/2012 36988.49 true 1 1 17/02/2012 54060.1 false 1 1 24/02/2012 20124.22 false 1 1 02/03/2012 20113.03 false 1 1 09/03/2012 21140.07 false 1 1 16/03/2012 22366.88 false 1 1 23/03/2012 22107.7 false 1 1 30/03/2012 28952.86 false 1 1 06/04/2012 57592.12 false 1 1 13/04/2012 34684.21 false 1 1 20/04/2012 16976.19 false 1 1 27/04/2012 16347.6 false 1 1 04/05/2012 17147.44 false 1 1 11/05/2012 18164.2 false 1 1 18/05/2012 18517.79 false 1 1 25/05/2012 16963.55 false 1 1 01/06/2012 16065.49 false 1 1 08/06/2012 17666.0 false 1 1 15/06/2012 17558.82 false 1 1 22/06/2012 16633.41 false 1 1 29/06/2012 15722.82 false 1 1 06/07/2012 17823.37 false 1 1 13/07/2012 16566.18 false 1 1 20/07/2012 16348.06 false 1 1 27/07/2012 15731.18 false 1 1 03/08/2012 16628.31 false 1 1 10/08/2012 16119.92 false 1 1 17/08/2012 17330.7 false 

In [0]:
sales.printSchema()

root
-- Store: integer (nullable = true)
-- Dept: integer (nullable = true)
-- Date: string (nullable = true)
-- Weekly_Sales: double (nullable = true)
-- IsHoliday: boolean (nullable = true)

Se observa que Date está mal inferido, por lo que se procede a declarar el esquema.

In [0]:
salesSchema = StructType([
  StructField("Store", IntegerType(), True), 
  StructField("Dept", IntegerType(), True),
  StructField("Date", StringType(), True),
  StructField("Weekly_Sales", DoubleType(), True),
  StructField("IsHoliday", BooleanType(), True)
])

sales2 = (spark.read
          .format(file_type)
          .option("header", first_row_is_header)
          .option("sep", delimiter)
          .schema(salesSchema)
          .load(file_location))

display(sales2)

Store Dept Date Weekly_Sales IsHoliday 1 1 05/02/2010 24924.5 false 1 1 12/02/2010 46039.49 true 1 1 19/02/2010 41595.55 false 1 1 26/02/2010 19403.54 false 1 1 05/03/2010 21827.9 false 1 1 12/03/2010 21043.39 false 1 1 19/03/2010 22136.64 false 1 1 26/03/2010 26229.21 false 1 1 02/04/2010 57258.43 false 1 1 09/04/2010 42960.91 false 1 1 16/04/2010 17596.96 false 1 1 23/04/2010 16145.35 false 1 1 30/04/2010 16555.11 false 1 1 07/05/2010 17413.94 false 1 1 14/05/2010 18926.74 false 1 1 21/05/2010 14773.04 false 1 1 28/05/2010 15580.43 false 1 1 04/06/2010 17558.09 false 1 1 11/06/2010 16637.62 false 1 1 18/06/2010 16216.27 false 1 1 25/06/2010 16328.72 false 1 1 02/07/2010 16333.14 false 1 1 09/07/2010 17688.76 false 1 1 16/07/2010 17150.84 false 1 1 23/07/2010 15360.45 false 1 1 30/07/2010 15381.82 false 1 1 06/08/2010 17508.41 false 1 1 13/08/2010 15536.4 false 1 1 20/08/2010 15740.13 false 1 1 27/08/2010 15793.87 false 1 1 03/09/2010 16241.78 false 1 1 10/09/2010 18194.74 true 1 1 17/09/2010 19354.23 false 1 1 24/09/2010 18122.52 false 1 1 01/10/2010 20094.19 false 1 1 08/10/2010 23388.03 false 1 1 15/10/2010 26978.34 false 1 1 22/10/2010 25543.04 false 1 1 29/10/2010 38640.93 false 1 1 05/11/2010 34238.88 false 1 1 12/11/2010 19549.39 false 1 1 19/11/2010 19552.84 false 1 1 26/11/2010 18820.29 true 1 1 03/12/2010 22517.56 false 1 1 10/12/2010 31497.65 false 1 1 17/12/2010 44912.86 false 1 1 24/12/2010 55931.23 false 1 1 31/12/2010 19124.58 true 1 1 07/01/2011 15984.24 false 1 1 14/01/2011 17359.7 false 1 1 21/01/2011 17341.47 false 1 1 28/01/2011 18461.18 false 1 1 04/02/2011 21665.76 false 1 1 11/02/2011 37887.17 true 1 1 18/02/2011 46845.87 false 1 1 25/02/2011 19363.83 false 1 1 04/03/2011 20327.61 false 1 1 11/03/2011 21280.4 false 1 1 18/03/2011 20334.23 false 1 1 25/03/2011 20881.1 false 1 1 01/04/2011 20398.09 false 1 1 08/04/2011 23873.79 false 1 1 15/04/2011 28762.37 false 1 1 22/04/2011 50510.31 false 1 1 29/04/2011 41512.39 false 1 1 06/05/2011 20138.19 false 1 1 13/05/2011 17235.15 false 1 1 20/05/2011 15136.78 false 1 1 27/05/2011 15741.6 false 1 1 03/06/2011 16434.15 false 1 1 10/06/2011 15883.52 false 1 1 17/06/2011 14978.09 false 1 1 24/06/2011 15682.81 false 1 1 01/07/2011 15363.5 false 1 1 08/07/2011 16148.87 false 1 1 15/07/2011 15654.85 false 1 1 22/07/2011 15766.6 false 1 1 29/07/2011 15922.41 false 1 1 05/08/2011 15295.55 false 1 1 12/08/2011 14539.79 false 1 1 19/08/2011 14689.24 false 1 1 26/08/2011 14537.37 false 1 1 02/09/2011 15277.27 false 1 1 09/09/2011 17746.68 true 1 1 16/09/2011 18535.48 false 1 1 23/09/2011 17859.3 false 1 1 30/09/2011 18337.68 false 1 1 07/10/2011 20797.58 false 1 1 14/10/2011 23077.55 false 1 1 21/10/2011 23351.8 false 1 1 28/10/2011 31579.9 false 1 1 04/11/2011 39886.06 false 1 1 11/11/2011 18689.54 false 1 1 18/11/2011 19050.66 false 1 1 25/11/2011 20911.25 true 1 1 02/12/2011 25293.49 false 1 1 09/12/2011 33305.92 false 1 1 16/12/2011 45773.03 false 1 1 23/12/2011 46788.75 false 1 1 30/12/2011 23350.88 true 1 1 06/01/2012 16567.69 false 1 1 13/01/2012 16894.4 false 1 1 20/01/2012 18365.1 false 1 1 27/01/2012 18378.16 false 1 1 03/02/2012 23510.49 false 1 1 10/02/2012 36988.49 true 1 1 17/02/2012 54060.1 false 1 1 24/02/2012 20124.22 false 1 1 02/03/2012 20113.03 false 1 1 09/03/2012 21140.07 false 1 1 16/03/2012 22366.88 false 1 1 23/03/2012 22107.7 false 1 1 30/03/2012 28952.86 false 1 1 06/04/2012 57592.12 false 1 1 13/04/2012 34684.21 false 1 1 20/04/2012 16976.19 false 1 1 27/04/2012 16347.6 false 1 1 04/05/2012 17147.44 false 1 1 11/05/2012 18164.2 false 1 1 18/05/2012 18517.79 false 1 1 25/05/2012 16963.55 false 1 1 01/06/2012 16065.49 false 1 1 08/06/2012 17666.0 false 1 1 15/06/2012 17558.82 false 1 1 22/06/2012 16633.41 false 1 1 29/06/2012 15722.82 false 1 1 06/07/2012 17823.37 false 1 1 13/07/2012 16566.18 false 1 1 20/07/2012 16348.06 false 1 1 27/07/2012 15731.18 false 1 1 03/08/2012 16628.31 false 1 1 10/08/2012 16119.92 false 1 1 17/08/2012 17330.7 false 

Cambiando el tipo de Date (String -> Date)

In [0]:
sales3 = sales2.withColumn("Date", to_date(col("Date"),"dd/MM/yyyy"))
#changedTypedf = joindf.withColumn("show", joindf["show"].cast(DoubleType()))
display(sales3)

Store Dept Date Weekly_Sales IsHoliday 1 1 2010-02-05 24924.5 false 1 1 2010-02-12 46039.49 true 1 1 2010-02-19 41595.55 false 1 1 2010-02-26 19403.54 false 1 1 2010-03-05 21827.9 false 1 1 2010-03-12 21043.39 false 1 1 2010-03-19 22136.64 false 1 1 2010-03-26 26229.21 false 1 1 2010-04-02 57258.43 false 1 1 2010-04-09 42960.91 false 1 1 2010-04-16 17596.96 false 1 1 2010-04-23 16145.35 false 1 1 2010-04-30 16555.11 false 1 1 2010-05-07 17413.94 false 1 1 2010-05-14 18926.74 false 1 1 2010-05-21 14773.04 false 1 1 2010-05-28 15580.43 false 1 1 2010-06-04 17558.09 false 1 1 2010-06-11 16637.62 false 1 1 2010-06-18 16216.27 false 1 1 2010-06-25 16328.72 false 1 1 2010-07-02 16333.14 false 1 1 2010-07-09 17688.76 false 1 1 2010-07-16 17150.84 false 1 1 2010-07-23 15360.45 false 1 1 2010-07-30 15381.82 false 1 1 2010-08-06 17508.41 false 1 1 2010-08-13 15536.4 false 1 1 2010-08-20 15740.13 false 1 1 2010-08-27 15793.87 false 1 1 2010-09-03 16241.78 false 1 1 2010-09-10 18194.74 true 1 1 2010-09-17 19354.23 false 1 1 2010-09-24 18122.52 false 1 1 2010-10-01 20094.19 false 1 1 2010-10-08 23388.03 false 1 1 2010-10-15 26978.34 false 1 1 2010-10-22 25543.04 false 1 1 2010-10-29 38640.93 false 1 1 2010-11-05 34238.88 false 1 1 2010-11-12 19549.39 false 1 1 2010-11-19 19552.84 false 1 1 2010-11-26 18820.29 true 1 1 2010-12-03 22517.56 false 1 1 2010-12-10 31497.65 false 1 1 2010-12-17 44912.86 false 1 1 2010-12-24 55931.23 false 1 1 2010-12-31 19124.58 true 1 1 2011-01-07 15984.24 false 1 1 2011-01-14 17359.7 false 1 1 2011-01-21 17341.47 false 1 1 2011-01-28 18461.18 false 1 1 2011-02-04 21665.76 false 1 1 2011-02-11 37887.17 true 1 1 2011-02-18 46845.87 false 1 1 2011-02-25 19363.83 false 1 1 2011-03-04 20327.61 false 1 1 2011-03-11 21280.4 false 1 1 2011-03-18 20334.23 false 1 1 2011-03-25 20881.1 false 1 1 2011-04-01 20398.09 false 1 1 2011-04-08 23873.79 false 1 1 2011-04-15 28762.37 false 1 1 2011-04-22 50510.31 false 1 1 2011-04-29 41512.39 false 1 1 2011-05-06 20138.19 false 1 1 2011-05-13 17235.15 false 1 1 2011-05-20 15136.78 false 1 1 2011-05-27 15741.6 false 1 1 2011-06-03 16434.15 false 1 1 2011-06-10 15883.52 false 1 1 2011-06-17 14978.09 false 1 1 2011-06-24 15682.81 false 1 1 2011-07-01 15363.5 false 1 1 2011-07-08 16148.87 false 1 1 2011-07-15 15654.85 false 1 1 2011-07-22 15766.6 false 1 1 2011-07-29 15922.41 false 1 1 2011-08-05 15295.55 false 1 1 2011-08-12 14539.79 false 1 1 2011-08-19 14689.24 false 1 1 2011-08-26 14537.37 false 1 1 2011-09-02 15277.27 false 1 1 2011-09-09 17746.68 true 1 1 2011-09-16 18535.48 false 1 1 2011-09-23 17859.3 false 1 1 2011-09-30 18337.68 false 1 1 2011-10-07 20797.58 false 1 1 2011-10-14 23077.55 false 1 1 2011-10-21 23351.8 false 1 1 2011-10-28 31579.9 false 1 1 2011-11-04 39886.06 false 1 1 2011-11-11 18689.54 false 1 1 2011-11-18 19050.66 false 1 1 2011-11-25 20911.25 true 1 1 2011-12-02 25293.49 false 1 1 2011-12-09 33305.92 false 1 1 2011-12-16 45773.03 false 1 1 2011-12-23 46788.75 false 1 1 2011-12-30 23350.88 true 1 1 2012-01-06 16567.69 false 1 1 2012-01-13 16894.4 false 1 1 2012-01-20 18365.1 false 1 1 2012-01-27 18378.16 false 1 1 2012-02-03 23510.49 false 1 1 2012-02-10 36988.49 true 1 1 2012-02-17 54060.1 false 1 1 2012-02-24 20124.22 false 1 1 2012-03-02 20113.03 false 1 1 2012-03-09 21140.07 false 1 1 2012-03-16 22366.88 false 1 1 2012-03-23 22107.7 false 1 1 2012-03-30 28952.86 false 1 1 2012-04-06 57592.12 false 1 1 2012-04-13 34684.21 false 1 1 2012-04-20 16976.19 false 1 1 2012-04-27 16347.6 false 1 1 2012-05-04 17147.44 false 1 1 2012-05-11 18164.2 false 1 1 2012-05-18 18517.79 false 1 1 2012-05-25 16963.55 false 1 1 2012-06-01 16065.49 false 1 1 2012-06-08 17666.0 false 1 1 2012-06-15 17558.82 false 1 1 2012-06-22 16633.41 false 1 1 2012-06-29 15722.82 false 1 1 2012-07-06 17823.37 false 1 1 2012-07-13 16566.18 false 1 1 2012-07-20 16348.06 false 1 1 2012-07-27 15731.18 false 1 1 2012-08-03 16628.31 false 1 1 2012-08-10 16119.92 false 1 1 2012-08-17 17330.7 false 

In [0]:
sales3.printSchema()

root
-- Store: integer (nullable = true)
-- Dept: integer (nullable = true)
-- Date: date (nullable = true)
-- Weekly_Sales: double (nullable = true)
-- IsHoliday: boolean (nullable = true)

In [0]:
sales3.count()

Out[15]: 421570

### stores data-set.csv

In [0]:
%fs head FileStore/tables/stores_data_set.csv

Store,Type,Size
1,A,151315
2,A,202307
3,B,37392
4,A,205863
5,B,34875
6,A,202505
7,B,70713
8,A,155078
9,B,125833
10,B,126512
11,A,207499
12,B,112238
13,A,219622
14,A,200898
15,B,123737
16,B,57197
17,B,93188
18,B,120653
19,A,203819
20,A,203742
21,B,140167
22,B,119557
23,B,114533
24,A,203819
25,B,128107
26,A,152513
27,A,204184
28,A,206302
29,B,93638
30,C,42988
31,A,203750
32,A,203007
33,A,39690
34,A,158114
35,B,103681
36,A,39910
37,C,39910
38,C,39690
39,A,184109
40,A,155083
41,A,196321
42,C,39690
43,C,41062
44,C,39910
45,B,118221

In [0]:
# File location and type
file_location = "/FileStore/tables/stores_data_set.csv"
file_type = "csv"

# CSV options
infer_schema = "true"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
stores = (spark.read.format(file_type)
          .option("inferSchema", infer_schema)
          .option("header", first_row_is_header)
          .option("sep", delimiter)
          .load(file_location))

display(stores)

Store Type Size 1 A 151315 2 A 202307 3 B 37392 4 A 205863 5 B 34875 6 A 202505 7 B 70713 8 A 155078 9 B 125833 10 B 126512 11 A 207499 12 B 112238 13 A 219622 14 A 200898 15 B 123737 16 B 57197 17 B 93188 18 B 120653 19 A 203819 20 A 203742 21 B 140167 22 B 119557 23 B 114533 24 A 203819 25 B 128107 26 A 152513 27 A 204184 28 A 206302 29 B 93638 30 C 42988 31 A 203750 32 A 203007 33 A 39690 34 A 158114 35 B 103681 36 A 39910 37 C 39910 38 C 39690 39 A 184109 40 A 155083 41 A 196321 42 C 39690 43 C 41062 44 C 39910 45 B 118221

In [0]:
stores.printSchema()

root
-- Store: integer (nullable = true)
-- Type: string (nullable = true)
-- Size: integer (nullable = true)

Se observa que se inferió correctamente el esquema.

In [0]:
stores.count()

Out[18]: 45

## Joinning datasets

In [0]:
# features3, sales3, stores
featuresSales = features3.join(sales3, on=['Store','Date','IsHoliday'], how='full') # Store, Date, IsHoliday

display(featuresSales)

Store Date IsHoliday Temperature Fuel_Price MarkDown1 MarkDown2 MarkDown3 MarkDown4 MarkDown5 CPI Unemployment Dept Weekly_Sales 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 1 14539.79 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 2 45341.92 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 3 32148.93 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 4 37449.94 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 5 21320.93 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 6 3124.83 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 7 14088.28 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 8 31061.21 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 9 18708.23 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 10 31383.62 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 11 23696.66 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 12 11581.03 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 13 37593.97 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 14 12652.58 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 16 17998.16 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 17 8893.77 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 19 876.41 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 20 4242.47 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 21 7662.32 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 22 8233.21 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 23 21142.51 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 24 7863.47 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 25 8731.19 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 26 4529.47 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 27 815.45 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 28 302.76 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 29 3339.78 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 30 3114.22 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 31 4521.44 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 32 5605.12 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 33 7364.64 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 34 11185.39 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 35 1806.76 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 36 423.0 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 37 2620.3 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 38 69572.84 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 40 51106.65 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 41 969.5 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 42 7308.1 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 44 4532.06 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 46 17231.34 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 47 0.0 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 48 519.0 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 49 14666.76 1 2011-08-12 false 90.76 3.638 null null null null null 215.6057878 7.962 52 2572.2 1 2011-08-12 false 90.76 3.638 null null null null 

In [0]:
featuresSales.count()

Out[20]: 423325

Dados los conteos de los registros de las tablas features3 (8,190) y sales3 (421,570), se observa que al hacer outerjoin de estas tablas, se obtienen (423,325) registros, lo que indica que hay registros en cada tabla que no tengan el id correspondiente entre sí.

In [0]:
featuresSalesStores = featuresSales.join(stores, on=['Store'], how='full') # Store

display(featuresSalesStores)

Store Date IsHoliday Temperature Fuel_Price MarkDown1 MarkDown2 MarkDown3 MarkDown4 MarkDown5 CPI Unemployment Dept Weekly_Sales Type Size 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 1 15169.49 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 2 58356.34 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 3 7289.98 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 4 31998.55 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 5 20564.76 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 6 3020.31 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 7 22689.32 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 8 26618.32 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 9 14553.3 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 10 19984.26 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 11 11672.42 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 12 2879.44 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 13 40461.94 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 14 16163.86 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 16 16036.04 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 17 7075.39 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 19 32.0 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 20 7363.16 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 21 5308.91 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 22 8708.72 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 23 24670.87 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 24 4470.96 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 25 10083.75 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 26 7201.78 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 27 788.1 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 28 373.71 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 29 4439.36 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 30 3920.84 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 31 1243.92 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 32 6904.82 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 33 5024.78 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 34 16090.8 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 35 2345.0 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 36 3011.0 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 37 3316.58 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 38 77099.36 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 40 51989.34 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 41 609.0 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 42 4931.23 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 44 2924.84 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 45 -4.72 A 203750 31 2010-06-18 false 86.18 2.637 null null null null nu

In [0]:
featuresSalesStores.count()

Out[22]: 423325

## Explorando data

### Mostrando tabla

In [0]:
display(featuresSalesStores)

Store Date IsHoliday Temperature Fuel_Price MarkDown1 MarkDown2 MarkDown3 MarkDown4 MarkDown5 CPI Unemployment Dept Weekly_Sales Type Size 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 1 15169.49 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 2 58356.34 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 3 7289.98 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 4 31998.55 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 5 20564.76 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 6 3020.31 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 7 22689.32 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 8 26618.32 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 9 14553.3 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 10 19984.26 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 11 11672.42 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 12 2879.44 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 13 40461.94 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 14 16163.86 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 16 16036.04 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 17 7075.39 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 19 32.0 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 20 7363.16 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 21 5308.91 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 22 8708.72 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 23 24670.87 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 24 4470.96 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 25 10083.75 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 26 7201.78 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 27 788.1 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 28 373.71 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 29 4439.36 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 30 3920.84 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 31 1243.92 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 32 6904.82 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 33 5024.78 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 34 16090.8 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 35 2345.0 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 36 3011.0 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 37 3316.58 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 38 77099.36 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 40 51989.34 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 41 609.0 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 42 4931.23 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 44 2924.84 A 203750 31 2010-06-18 false 86.18 2.637 null null null null null 211.1096543 8.2 45 -4.72 A 203750 31 2010-06-18 false 86.18 2.637 null null null null nu

### Análisis estadístico descriptivo

In [0]:
featuresSalesStores.describe().display()

summary Store Temperature Fuel_Price MarkDown1 MarkDown2 MarkDown3 MarkDown4 MarkDown5 CPI Unemployment Dept Weekly_Sales Type Size count 423325 423325 423325 152433 112532 138658 136466 153187 422740 422740 421570 421570 423325 423325 mean 22.20385991850233 60.06713962085865 3.361933245142616 7246.604247374274 3337.5972891266547 1449.0989315437912 3382.0198340978786 4618.743782827524 171.21886315838344 7.957296073236492 44.26031738501317 15981.258123467032 null 136701.2157940117 stddev 12.78624408303602 18.452598879854165 0.45803810684327795 8319.906707191132 9461.265420282907 9674.70523428954 6306.728467945709 6258.233495654179 39.16708815989315 1.8634333014166284 30.492054015785836 22711.18351916323 null 60990.97780085184 min 1 -7.29 2.472 -2781.45 -265.76 -179.26 0.22 -185.17 126.064 3.684 1 -4988.94 A 34875 max 45 101.95 4.468 103184.98 104519.54 149483.31 67474.85 771448.1 228.9764563 14.313 99 693099.36 C 219622

### Realizando algunas agrupaciones

#### Groupby type

In [0]:
(featuresSalesStores
 .groupBy("Type")
 .agg(sf.sum('Weekly_Sales').alias('Weekly_Sales_Sum'))
 .withColumn("Weekly_Sales_Sum", sf.round(col("Weekly_Sales_Sum"), 2))
 .sort('Type')
 .display())

Type Weekly_Sales_Sum A 4.33101472275E9 B 2.00070073682E9 C 4.0550352754E8

In [0]:
# featuresSalesStores.groupBy("type").agg({'Weekly_Sales':'avg'}).display()
(featuresSalesStores
 .groupBy("Type")
 .agg(sf.avg('Weekly_Sales').alias('Weekly_Sales_Avg'))
 .withColumn("Weekly_Sales_Avg", sf.round(col("Weekly_Sales_Avg"), 2))
 .sort('Type')
 .display())

Type Weekly_Sales_Avg A 20099.57 B 12237.08 C 9519.53

In [0]:
(featuresSalesStores
 .groupBy("Type")
 .count()
 .sort('Type')
 .display())

Type count A 216336 B 164158 C 42831

#### Groupby type & Date

In [0]:
(featuresSalesStores
 .groupBy(['Date','Type'])
 .agg(sf.sum('Weekly_Sales').alias('Weekly_Sales_Sum'))
 .withColumn('Weekly_Sales_Sum', sf.round(col('Weekly_Sales_Sum'), 2))
 .sort('Date','Type')
 .display())

Date Type Weekly_Sales_Sum 2010-02-05 A 3.214412625E7 2010-02-05 B 1.477549862E7 2010-02-05 C 2831115.63 2010-02-12 A 3.098257079E7 2010-02-12 B 1.443917859E7 2010-02-12 C 2914928.25 2010-02-19 A 3.100007276E7 2010-02-19 B 1.454003862E7 2010-02-19 C 2736882.4 2010-02-26 A 2.800899599E7 2010-02-26 B 1.325611344E7 2010-02-26 C 2703461.7 2010-03-05 A 3.00205436E7 2010-03-05 B 1.398865032E7 2010-03-05 C 2862276.38 2010-03-12 A 2.93775864E7 2010-03-12 B 1.371863952E7 2010-03-12 C 2829170.59 2010-03-19 A 2.87485756E7 2010-03-19 B 1.354397968E7 2010-03-19 C 2696419.36 2010-03-26 A 2.787290822E7 2010-03-26 B 1.35895603E7 2010-03-26 C 2671492.53 2010-04-02 A 3.231494992E7 2010-04-02 B 1.529967032E7 2010-04-02 C 2809211.02 2010-04-09 A 3.031617763E7 2010-04-09 B 1.420638194E7 2010-04-09 C 2842730.87 2010-04-16 A 2.893962379E7 2010-04-16 B 1.340586171E7 2010-04-16 C 2838181.58 2010-04-23 A 2.846722192E7 2010-04-23 B 1.353302549E7 2010-04-23 C 2734205.15 2010-04-30 A 2.795921128E7 2010-04-30 B 1.305259098E7 2010-04-30 C 2693324.45 2010-05-07 A 3.13582192E7 2010-05-07 B 1.420052349E7 2010-05-07 C 2944500.83 2010-05-14 A 2.917231559E7 2010-05-14 B 1.330281308E7 2010-05-14 C 2854951.53 2010-05-21 A 2.894388063E7 2010-05-21 B 1.342595322E7 2010-05-21 C 2750274.21 2010-05-28 A 3.033060158E7 2010-05-28 B 1.469941237E7 2010-05-28 C 2727488.61 2010-06-04 A 3.229781555E7 2010-06-04 B 1.50565935E7 2010-06-04 C 2834134.07 2010-06-11 A 3.074807111E7 2010-06-11 B 1.420814094E7 2010-06-11 C 2870334.67 2010-06-18 A 3.039227464E7 2010-06-18 B 1.441768083E7 2010-06-18 C 2812090.76 2010-06-25 A 2.954936265E7 2010-06-25 B 1.440122631E7 2010-06-25 C 2658447.33 2010-07-02 A 3.123864048E7 2010-07-02 B 1.489388176E7 2010-07-02 C 2784962.26 2010-07-09 A 3.07366697E7 2010-07-09 B 1.43109693E7 2010-07-09 C 2851890.37 2010-07-16 A 2.956461638E7 2010-07-16 B 1.386500519E7 2010-07-16 C 2814278.01 2010-07-23 A 2.859555431E7 2010-07-23 B 1.359963693E7 2010-07-23 C 2693657.87 2010-07-30 A 2.84177837E7 2010-07-30 B 1.358366287E7 2010-07-30 C 2628916.85 2010-08-06 A 3.101181662E7 2010-08-06 B 1.428973962E7 2010-08-06 C 2903442.88 2010-08-13 A 2.986389792E7 2010-08-13 B 1.37355559E7 2010-08-13 C 2864964.59 2010-08-20 A 3.025829208E7 2010-08-20 B 1.404089537E7 2010-08-20 C 2761765.47 2010-08-27 A 2.910635527E7 2010-08-27 B 1.41101153E7 2010-08-27 C 2693269.87 2010-09-03 A 3.034546892E7 2010-09-03 B 1.398985054E7 2010-09-03 C 2858938.15 2010-09-10 A 2.934210237E7 2010-09-10 B 1.339594258E7 2010-09-10 C 2896352.89 2010-09-17 A 2.769430996E7 2010-09-17 B 1.254557502E7 2010-09-17 C 2840841.53 2010-09-24 A 2.651819152E7 2010-09-24 B 1.214321965E7 2010-09-24 C 2697103.24 2010-10-01 A 2.726542416E7 2010-10-01 B 1.220005675E7 2010-10-01 C 2774394.96 2010-10-08 A 2.914119104E7 2010-10-08 B 1.303133128E7 2010-10-08 C 2930451.91 2010-10-15 A 2.794287226E7 2010-10-15 B 1.238814172E7 2010-10-15 C 2818459.47 2010-10-22 A 2.755980574E7 2010-10-22 B 1.272424362E7 2010-10-22 C 2782620.16 2010-10-29 A 2.798131834E7 2010-10-29 B 1.295393123E7 2010-10-29 C 2667581.38 2010-11-05 A 2.970328648E7 2010-11-05 B 1.323161259E7 2010-11-05 C 2847082.82 2010-11-12 A 2.976287873E7 2010-11-12 B 1.360703794E7 2010-11-12 C 2754884.16 2010-11-19 A 2.917204008E7 2010-11-19 B 1.326501218E7 2010-11-19 C 2688531.92 2010-11-26 A 4.144845721E7 2010-11-26 B 2.156727142E7 2010-11-26 C 2805274.61 2010-12-03 A 3.157629925E7 2010-12-03 B 1.574346937E7 2010-12-03 C 2589259.26 2010-12-10 A 3.556781725E7 2010-12-10 B 1.734104094E7 2010-12-10 C 2757912.2 2010-12-17 A 3.917942301E7 2010-12-17 B 1.989545443E7 2010-12-17 C 2745922.41 2010-12-24 A 5.1499373E7 2010-12-24 B 2.638995994E7 2010-12-24 C 3042082.66 2010-12-31 A 2.599142726E7 2010-12-31 B 1.20735892E7 2010-12-31 C 2367502.54 2011-01-07 A 2.739203248E7 2011-01-07 B 1.251080478E7 2011-01-07 C 2872950.51 2011-01-14 A 2.633331826E7 2011-01-14 B 1.147149679E7 2011-01-14 C 2868862.99 2011-01-21 A 2.631172442E7 2011-01-21 B 1.162562888E7 2011-01-21 C 2717294.73 2011-01-28 A

#### Groupby type & dept

In [0]:
(featuresSalesStores
 .groupBy(['Dept','Type'])
 .agg(sf.sum('Weekly_Sales').alias('Weekly_Sales_Sum'))
 .withColumn('Weekly_Sales_Sum', sf.round(col('Weekly_Sales_Sum'), 2))
 .sort('Dept','Type')
 .display())

Dept Type Weekly_Sales_Sum null A null null B null null C null 1 A 7.222236929E7 1 B 4.373581994E7 1 C 7680587.31 2 A 1.6357524715E8 2 B 1.0465940451E8 2 C 1.237652277E7 3 A 4.366973025E7 3 B 3.151892219E7 3 C 703797.51 4 A 1.0373561908E8 4 B 5.16828067E7 4 C 1.17283198E7 5 A 8.349274066E7 5 B 5.149976969E7 5 C 614848.22 6 A 1.622523174E7 6 B 1.2171675E7 6 C 23760.4 7 A 9.68539264E7 7 B 5.813915078E7 7 C 484485.57 8 A 1.1668898125E8 8 B 6.704432637E7 8 C 1.054747311E7 9 A 7.36850109E7 9 B 5.460999902E7 9 C 98246.73 10 A 7.474245456E7 10 B 4.286679633E7 10 C 288136.69 11 A 5.394099504E7 11 B 3.860628128E7 11 C 782000.06 12 A 1.614522698E7 12 B 1.053313971E7 12 C 81752.82 13 A 1.1772633169E8 13 B 6.838738753E7 13 C 1.120785073E7 14 A 5.580626862E7 14 B 3.884682773E7 14 C 1041570.1 16 A 5.779447148E7 16 B 3.253900507E7 16 C 1337205.72 17 A 3.732154782E7 17 B 2.766216356E7 17 C 336105.25 18 A 2.296333025E7 18 B 1.366894903E7 18 C 265209.78 19 A 3469705.74 19 B 3346474.99 19 C 2.38 20 A 1.901842015E7 20 B 1.430807575E7 20 C 23149.21 21 A 2.057120202E7 21 B 1.032954874E7 21 C 1163894.32 22 A 3.447759167E7 22 B 2.523378964E7 22 C 11726.05 23 A 7.491576205E7 23 B 6.324977327E7 23 C 14795.46 24 A 1.884461811E7 24 B 1.550213599E7 24 C 832.16 25 A 3.199300275E7 25 B 2.506893743E7 25 C 49278.44 26 A 2.573251483E7 26 B 1.816270867E7 26 C 4786.1 27 A 4951289.29 27 B 3939600.27 27 C 6447.03 28 A 2017700.11 28 B 1692640.47 28 C 74812.67 29 A 1.659251072E7 29 B 1.341688037E7 29 C 211.33 30 A 1.210604884E7 30 B 9683332.59 31 A 7655430.19 31 B 6497075.77 31 C 40878.26 32 A 2.380131529E7 32 B 1.687269901E7 32 C 24946.51 33 A 1.975666843E7 33 B 1.512678584E7 33 C 515.66 34 A 4.547261293E7 34 B 3.775079989E7 34 C 212.61 35 A 8691837.75 35 B 6780930.33 35 C 7.0 36 A 5916290.51 36 B 4793223.26 37 A 7011917.15 37 B 1005326.2 38 A 2.0450578691E8 38 B 1.4173013872E8 38 C 4.688221129E7 39 A 149.98 39 B 28.0 40 A 1.6876787976E8 40 B 1.0299066789E8 40 C 1.71774744E7 41 A 5767907.13 41 B 4830316.44 41 C 2041.5 42 A 1.904506622E7 42 B 1.407627022E7 42 C 97955.13 43 A 2.82 43 B 11.5 44 A 1.485468094E7 44 B 1.059421723E7 44 C 714.79 45 A 23736.75 45 B 21200.88 46 A 7.428580326E7 46 B 4.844629126E7 46 C 5612316.92 47 A 1926.03 47 B -6888.96 48 A 1655952.23 48 B 686852.38 49 A 2.405005948E7 49 B 1.251989444E7 49 C 2947.79 50 A 1810788.65 50 B 2342408.48 51 A 17517.64 51 B 13055.19 52 A 6893511.96 52 B 5168910.56 52 C 24514.47 54 A 295186.32 54 B 221108.31 55 A 3.697541608E7 55 B 2.299467679E7 55 C 362.48 56 A 1.370947926E7 56 B 8651962.75 56 C 150080.86 58 A 1.243134563E7 58 B 3824417.94 59 A 2578724.02 59 B 1581278.26 59 C 80392.24 60 A 1093649.24 60 B 485964.12 60 C 425407.6 65 A 6498163.99 67 A 3.009097897E7 67 B 1.661155291E7 67 C 1178245.8 71 A 1.713374743E7 71 B 1.324979311E7 71 C -517.65 72 A 1.7850824574E8 72 B 1.271272258E8 72 C 89680.67 74 A 5.048946627E7 74 B 3.790813388E7 74 C 384385.42 77 A 29999.75 77 B 19344.52 78 A 1093.47 78 B 621.24 79 A 9.038343555E7 79 B 4.33141133E7 79 C 7698995.29 80 A 6.157307644E7 80 B 793812.09 80 C 9407171.67 81 A 7.081085034E7 81 B 1.799026428E7 81 C 1.075091469E7 82 A 6.272985641E7 82 B 3.177765097E7 82 C 5039572.95 83 A 1.619277508E7 83 B 1488242.53 83 C 2307813.23 85 A 7841665.23 85 B 5796551.46 85 C 72479.52 87 A 5.551305621E7 87 B 2.979437379E7 87 C 452645.61 90 A 2.2195187982E8 90 B 3.16835616E7 90 C 3.743302226E7 91 A 1.690476016E8 91 B 2.145881295E7 91 C 2.627529118E7 92 A 3.5284554971E8 92 B 7.893503062E7 92 C 5.216276154E7 93 A 1.362109934E8 93 B 3044717.29 93 C 2.04429525E7 94 A 1.6065693021E8 94 B 2111682.06 94 C 2.714383806E7 95 A 3.0545780593E8 95 B 1.0041189393E8 95 C 4.345046266E7 96 A 5.428977401E7 96 B 6016892.33 96 C 1.352724982E7 97 A 6.950711714E7 97 B 8475427.03 97 C 1.151396773E7 98 A 3.454250175E7 98 B 584785.21 98 C 4701632.41 99 A 357666.3 99 B 308.6 99 C 174.95

### Plots

#### Type ~ Date

In [0]:
(featuresSalesStores
 .groupBy(['Date','Type'])
 .agg(sf.sum('Weekly_Sales').alias('Weekly_Sales_Sum'))
 .withColumn('Weekly_Sales_Sum', sf.round(col('Weekly_Sales_Sum'), 2))
 .sort('Date','Type')
 .display())

Date Type Weekly_Sales_Sum 2010-02-05 A 3.214412625E7 2010-02-05 B 1.477549862E7 2010-02-05 C 2831115.63 2010-02-12 A 3.098257079E7 2010-02-12 B 1.443917859E7 2010-02-12 C 2914928.25 2010-02-19 A 3.100007276E7 2010-02-19 B 1.454003862E7 2010-02-19 C 2736882.4 2010-02-26 A 2.800899599E7 2010-02-26 B 1.325611344E7 2010-02-26 C 2703461.7 2010-03-05 A 3.00205436E7 2010-03-05 B 1.398865032E7 2010-03-05 C 2862276.38 2010-03-12 A 2.93775864E7 2010-03-12 B 1.371863952E7 2010-03-12 C 2829170.59 2010-03-19 A 2.87485756E7 2010-03-19 B 1.354397968E7 2010-03-19 C 2696419.36 2010-03-26 A 2.787290822E7 2010-03-26 B 1.35895603E7 2010-03-26 C 2671492.53 2010-04-02 A 3.231494992E7 2010-04-02 B 1.529967032E7 2010-04-02 C 2809211.02 2010-04-09 A 3.031617763E7 2010-04-09 B 1.420638194E7 2010-04-09 C 2842730.87 2010-04-16 A 2.893962379E7 2010-04-16 B 1.340586171E7 2010-04-16 C 2838181.58 2010-04-23 A 2.846722192E7 2010-04-23 B 1.353302549E7 2010-04-23 C 2734205.15 2010-04-30 A 2.795921128E7 2010-04-30 B 1.305259098E7 2010-04-30 C 2693324.45 2010-05-07 A 3.13582192E7 2010-05-07 B 1.420052349E7 2010-05-07 C 2944500.83 2010-05-14 A 2.917231559E7 2010-05-14 B 1.330281308E7 2010-05-14 C 2854951.53 2010-05-21 A 2.894388063E7 2010-05-21 B 1.342595322E7 2010-05-21 C 2750274.21 2010-05-28 A 3.033060158E7 2010-05-28 B 1.469941237E7 2010-05-28 C 2727488.61 2010-06-04 A 3.229781555E7 2010-06-04 B 1.50565935E7 2010-06-04 C 2834134.07 2010-06-11 A 3.074807111E7 2010-06-11 B 1.420814094E7 2010-06-11 C 2870334.67 2010-06-18 A 3.039227464E7 2010-06-18 B 1.441768083E7 2010-06-18 C 2812090.76 2010-06-25 A 2.954936265E7 2010-06-25 B 1.440122631E7 2010-06-25 C 2658447.33 2010-07-02 A 3.123864048E7 2010-07-02 B 1.489388176E7 2010-07-02 C 2784962.26 2010-07-09 A 3.07366697E7 2010-07-09 B 1.43109693E7 2010-07-09 C 2851890.37 2010-07-16 A 2.956461638E7 2010-07-16 B 1.386500519E7 2010-07-16 C 2814278.01 2010-07-23 A 2.859555431E7 2010-07-23 B 1.359963693E7 2010-07-23 C 2693657.87 2010-07-30 A 2.84177837E7 2010-07-30 B 1.358366287E7 2010-07-30 C 2628916.85 2010-08-06 A 3.101181662E7 2010-08-06 B 1.428973962E7 2010-08-06 C 2903442.88 2010-08-13 A 2.986389792E7 2010-08-13 B 1.37355559E7 2010-08-13 C 2864964.59 2010-08-20 A 3.025829208E7 2010-08-20 B 1.404089537E7 2010-08-20 C 2761765.47 2010-08-27 A 2.910635527E7 2010-08-27 B 1.41101153E7 2010-08-27 C 2693269.87 2010-09-03 A 3.034546892E7 2010-09-03 B 1.398985054E7 2010-09-03 C 2858938.15 2010-09-10 A 2.934210237E7 2010-09-10 B 1.339594258E7 2010-09-10 C 2896352.89 2010-09-17 A 2.769430996E7 2010-09-17 B 1.254557502E7 2010-09-17 C 2840841.53 2010-09-24 A 2.651819152E7 2010-09-24 B 1.214321965E7 2010-09-24 C 2697103.24 2010-10-01 A 2.726542416E7 2010-10-01 B 1.220005675E7 2010-10-01 C 2774394.96 2010-10-08 A 2.914119104E7 2010-10-08 B 1.303133128E7 2010-10-08 C 2930451.91 2010-10-15 A 2.794287226E7 2010-10-15 B 1.238814172E7 2010-10-15 C 2818459.47 2010-10-22 A 2.755980574E7 2010-10-22 B 1.272424362E7 2010-10-22 C 2782620.16 2010-10-29 A 2.798131834E7 2010-10-29 B 1.295393123E7 2010-10-29 C 2667581.38 2010-11-05 A 2.970328648E7 2010-11-05 B 1.323161259E7 2010-11-05 C 2847082.82 2010-11-12 A 2.976287873E7 2010-11-12 B 1.360703794E7 2010-11-12 C 2754884.16 2010-11-19 A 2.917204008E7 2010-11-19 B 1.326501218E7 2010-11-19 C 2688531.92 2010-11-26 A 4.144845721E7 2010-11-26 B 2.156727142E7 2010-11-26 C 2805274.61 2010-12-03 A 3.157629925E7 2010-12-03 B 1.574346937E7 2010-12-03 C 2589259.26 2010-12-10 A 3.556781725E7 2010-12-10 B 1.734104094E7 2010-12-10 C 2757912.2 2010-12-17 A 3.917942301E7 2010-12-17 B 1.989545443E7 2010-12-17 C 2745922.41 2010-12-24 A 5.1499373E7 2010-12-24 B 2.638995994E7 2010-12-24 C 3042082.66 2010-12-31 A 2.599142726E7 2010-12-31 B 1.20735892E7 2010-12-31 C 2367502.54 2011-01-07 A 2.739203248E7 2011-01-07 B 1.251080478E7 2011-01-07 C 2872950.51 2011-01-14 A 2.633331826E7 2011-01-14 B 1.147149679E7 2011-01-14 C 2868862.99 2011-01-21 A 2.631172442E7 2011-01-21 B 1.162562888E7 2011-01-21 C 2717294.73 2011-01-28 A

#### Weekly Sales ~ Date

Analizando ventas semanales de departamentos. Se observa que las ventas del departamento 2 son constantes en el año y son las más altas, sin embargo los departamentos 4 y 7 son los que más venden en diciembre.

In [0]:
(featuresSalesStores
 .groupBy(['Date','Dept'])
 .agg(sf.sum('Weekly_Sales').alias('Weekly_Sales_Sum'))
 .withColumn('Weekly_Sales_Sum', sf.round(col('Weekly_Sales_Sum'), 2))
 .sort('Date','Dept')
 .display())

Date Dept Weekly_Sales_Sum 2010-02-05 1 881833.41 2010-02-05 2 1997831.89 2010-02-05 3 484368.9 2010-02-05 4 1205801.77 2010-02-05 5 1116952.54 2010-02-05 6 206793.28 2010-02-05 7 764967.96 2010-02-05 8 1527438.17 2010-02-05 9 535727.56 2010-02-05 10 773682.9 2010-02-05 11 651387.15 2010-02-05 12 168802.28 2010-02-05 13 1522916.77 2010-02-05 14 744664.42 2010-02-05 16 201198.58 2010-02-05 17 558684.58 2010-02-05 18 166777.43 2010-02-05 19 56493.12 2010-02-05 20 265131.96 2010-02-05 21 257986.69 2010-02-05 22 513801.98 2010-02-05 23 875453.87 2010-02-05 24 242250.33 2010-02-05 25 365598.14 2010-02-05 26 348794.25 2010-02-05 27 84463.38 2010-02-05 28 36758.58 2010-02-05 29 230289.15 2010-02-05 30 183381.81 2010-02-05 31 99402.34 2010-02-05 32 332272.16 2010-02-05 33 267483.5 2010-02-05 34 635170.08 2010-02-05 35 96689.31 2010-02-05 36 31678.9 2010-02-05 37 66989.64 2010-02-05 38 3022955.05 2010-02-05 40 2238389.82 2010-02-05 41 31182.71 2010-02-05 42 204211.2 2010-02-05 44 189772.1 2010-02-05 45 411.02 2010-02-05 46 854943.68 2010-02-05 47 202.0 2010-02-05 48 15809.96 2010-02-05 49 285793.0 2010-02-05 50 29780.0 2010-02-05 51 664.91 2010-02-05 52 99508.87 2010-02-05 54 9560.85 2010-02-05 55 555023.08 2010-02-05 56 39529.13 2010-02-05 58 155247.25 2010-02-05 59 34792.87 2010-02-05 60 14077.4 2010-02-05 65 41057.25 2010-02-05 67 365778.02 2010-02-05 71 345617.27 2010-02-05 72 3372690.55 2010-02-05 74 678429.57 2010-02-05 78 123.0 2010-02-05 79 1228686.53 2010-02-05 80 521864.34 2010-02-05 81 747310.8 2010-02-05 82 577369.2 2010-02-05 83 166944.95 2010-02-05 85 87214.76 2010-02-05 87 609066.67 2010-02-05 90 2199107.52 2010-02-05 91 1755822.24 2010-02-05 92 3781162.87 2010-02-05 93 1165221.8 2010-02-05 94 1301383.11 2010-02-05 95 3170529.93 2010-02-05 96 404386.06 2010-02-05 97 656075.0 2010-02-05 98 293125.35 2010-02-05 99 0.03 2010-02-12 1 1457182.4 2010-02-12 2 1839217.77 2010-02-12 3 429341.02 2010-02-12 4 1075081.62 2010-02-12 5 1148132.32 2010-02-12 6 252086.57 2010-02-12 7 781977.02 2010-02-12 8 1375373.29 2010-02-12 9 562174.72 2010-02-12 10 762338.46 2010-02-12 11 591729.99 2010-02-12 12 152456.1 2010-02-12 13 1327494.63 2010-02-12 14 714470.52 2010-02-12 16 252750.17 2010-02-12 17 536295.55 2010-02-12 18 512810.69 2010-02-12 19 52834.44 2010-02-12 20 241003.17 2010-02-12 21 252303.11 2010-02-12 22 501304.72 2010-02-12 23 895426.52 2010-02-12 24 244350.5 2010-02-12 25 394240.32 2010-02-12 26 340097.48 2010-02-12 27 78835.77 2010-02-12 28 33619.89 2010-02-12 29 251032.74 2010-02-12 30 173675.48 2010-02-12 31 98943.69 2010-02-12 32 428886.41 2010-02-12 33 257678.46 2010-02-12 34 611157.26 2010-02-12 35 86259.54 2010-02-12 36 33114.34 2010-02-12 37 62013.88 2010-02-12 38 2686547.52 2010-02-12 40 2037203.29 2010-02-12 41 37746.19 2010-02-12 42 184438.95 2010-02-12 43 0.25 2010-02-12 44 191664.18 2010-02-12 45 539.47 2010-02-12 46 805561.88 2010-02-12 47 1224.97 2010-02-12 48 12485.93 2010-02-12 49 291082.49 2010-02-12 50 30604.0 2010-02-12 51 885.67 2010-02-12 52 96251.15 2010-02-12 54 11502.94 2010-02-12 55 620547.99 2010-02-12 56 57485.76 2010-02-12 58 125943.61 2010-02-12 59 32788.75 2010-02-12 60 13909.2 2010-02-12 65 39799.36 2010-02-12 67 663589.77 2010-02-12 71 316181.96 2010-02-12 72 3102766.89 2010-02-12 74 607921.77 2010-02-12 78 93.0 2010-02-12 79 1082770.34 2010-02-12 80 516818.05 2010-02-12 81 680040.17 2010-02-12 82 591428.53 2010-02-12 83 171428.95 2010-02-12 85 88086.99 2010-02-12 87 541000.95 2010-02-12 90 2031157.15 2010-02-12 91 1603013.19 2010-02-12 92 3443837.3 2010-02-12 93 1077422.7 2010-02-12 94 1317069.86 2010-02-12 95 2976149.31 2010-02-12 96 448865.03 2010-02-12 97 613648.81 2010-02-12 98 377482.79 2010-02-12 99 0.01 2010-02-19 1 1118257.36 2010-02-19 2 1961685.93 2010-02-19 3 410399.14 2010-02-19 4 1113781.74 2010-02-19 5 1110717.47 2010-02-19 6 258741.79 2010-02-19 7 843175.36 2010-02-19 8 1414318.21 2010-02-19 9 587592.22 2010-02-19 10 780288.54 2010-02-19 11 582631.2 2010-02-19 12 165622.65 2010-0

#### Boxplot

Analizando las ventas semanales por cada departamento. Se observa que el departamento 16 tiene una gran varianza de ventas semanales.

In [0]:
(featuresSalesStores
 .groupBy(['Date','Dept'])
 .agg(sf.sum('Weekly_Sales').alias('Weekly_Sales_Sum'))
 .withColumn('Weekly_Sales_Sum', sf.round(col('Weekly_Sales_Sum'), 2))
 .sort('Date','Dept')
 .display())

Date Dept Weekly_Sales_Sum 2010-02-05 1 881833.41 2010-02-05 2 1997831.89 2010-02-05 3 484368.9 2010-02-05 4 1205801.77 2010-02-05 5 1116952.54 2010-02-05 6 206793.28 2010-02-05 7 764967.96 2010-02-05 8 1527438.17 2010-02-05 9 535727.56 2010-02-05 10 773682.9 2010-02-05 11 651387.15 2010-02-05 12 168802.28 2010-02-05 13 1522916.77 2010-02-05 14 744664.42 2010-02-05 16 201198.58 2010-02-05 17 558684.58 2010-02-05 18 166777.43 2010-02-05 19 56493.12 2010-02-05 20 265131.96 2010-02-05 21 257986.69 2010-02-05 22 513801.98 2010-02-05 23 875453.87 2010-02-05 24 242250.33 2010-02-05 25 365598.14 2010-02-05 26 348794.25 2010-02-05 27 84463.38 2010-02-05 28 36758.58 2010-02-05 29 230289.15 2010-02-05 30 183381.81 2010-02-05 31 99402.34 2010-02-05 32 332272.16 2010-02-05 33 267483.5 2010-02-05 34 635170.08 2010-02-05 35 96689.31 2010-02-05 36 31678.9 2010-02-05 37 66989.64 2010-02-05 38 3022955.05 2010-02-05 40 2238389.82 2010-02-05 41 31182.71 2010-02-05 42 204211.2 2010-02-05 44 189772.1 2010-02-05 45 411.02 2010-02-05 46 854943.68 2010-02-05 47 202.0 2010-02-05 48 15809.96 2010-02-05 49 285793.0 2010-02-05 50 29780.0 2010-02-05 51 664.91 2010-02-05 52 99508.87 2010-02-05 54 9560.85 2010-02-05 55 555023.08 2010-02-05 56 39529.13 2010-02-05 58 155247.25 2010-02-05 59 34792.87 2010-02-05 60 14077.4 2010-02-05 65 41057.25 2010-02-05 67 365778.02 2010-02-05 71 345617.27 2010-02-05 72 3372690.55 2010-02-05 74 678429.57 2010-02-05 78 123.0 2010-02-05 79 1228686.53 2010-02-05 80 521864.34 2010-02-05 81 747310.8 2010-02-05 82 577369.2 2010-02-05 83 166944.95 2010-02-05 85 87214.76 2010-02-05 87 609066.67 2010-02-05 90 2199107.52 2010-02-05 91 1755822.24 2010-02-05 92 3781162.87 2010-02-05 93 1165221.8 2010-02-05 94 1301383.11 2010-02-05 95 3170529.93 2010-02-05 96 404386.06 2010-02-05 97 656075.0 2010-02-05 98 293125.35 2010-02-05 99 0.03 2010-02-12 1 1457182.4 2010-02-12 2 1839217.77 2010-02-12 3 429341.02 2010-02-12 4 1075081.62 2010-02-12 5 1148132.32 2010-02-12 6 252086.57 2010-02-12 7 781977.02 2010-02-12 8 1375373.29 2010-02-12 9 562174.72 2010-02-12 10 762338.46 2010-02-12 11 591729.99 2010-02-12 12 152456.1 2010-02-12 13 1327494.63 2010-02-12 14 714470.52 2010-02-12 16 252750.17 2010-02-12 17 536295.55 2010-02-12 18 512810.69 2010-02-12 19 52834.44 2010-02-12 20 241003.17 2010-02-12 21 252303.11 2010-02-12 22 501304.72 2010-02-12 23 895426.52 2010-02-12 24 244350.5 2010-02-12 25 394240.32 2010-02-12 26 340097.48 2010-02-12 27 78835.77 2010-02-12 28 33619.89 2010-02-12 29 251032.74 2010-02-12 30 173675.48 2010-02-12 31 98943.69 2010-02-12 32 428886.41 2010-02-12 33 257678.46 2010-02-12 34 611157.26 2010-02-12 35 86259.54 2010-02-12 36 33114.34 2010-02-12 37 62013.88 2010-02-12 38 2686547.52 2010-02-12 40 2037203.29 2010-02-12 41 37746.19 2010-02-12 42 184438.95 2010-02-12 43 0.25 2010-02-12 44 191664.18 2010-02-12 45 539.47 2010-02-12 46 805561.88 2010-02-12 47 1224.97 2010-02-12 48 12485.93 2010-02-12 49 291082.49 2010-02-12 50 30604.0 2010-02-12 51 885.67 2010-02-12 52 96251.15 2010-02-12 54 11502.94 2010-02-12 55 620547.99 2010-02-12 56 57485.76 2010-02-12 58 125943.61 2010-02-12 59 32788.75 2010-02-12 60 13909.2 2010-02-12 65 39799.36 2010-02-12 67 663589.77 2010-02-12 71 316181.96 2010-02-12 72 3102766.89 2010-02-12 74 607921.77 2010-02-12 78 93.0 2010-02-12 79 1082770.34 2010-02-12 80 516818.05 2010-02-12 81 680040.17 2010-02-12 82 591428.53 2010-02-12 83 171428.95 2010-02-12 85 88086.99 2010-02-12 87 541000.95 2010-02-12 90 2031157.15 2010-02-12 91 1603013.19 2010-02-12 92 3443837.3 2010-02-12 93 1077422.7 2010-02-12 94 1317069.86 2010-02-12 95 2976149.31 2010-02-12 96 448865.03 2010-02-12 97 613648.81 2010-02-12 98 377482.79 2010-02-12 99 0.01 2010-02-19 1 1118257.36 2010-02-19 2 1961685.93 2010-02-19 3 410399.14 2010-02-19 4 1113781.74 2010-02-19 5 1110717.47 2010-02-19 6 258741.79 2010-02-19 7 843175.36 2010-02-19 8 1414318.21 2010-02-19 9 587592.22 2010-02-19 10 780288.54 2010-02-19 11 582631.2 2010-02-19 12 165622.65 2010-0

## RFM Análisis (Recency, Frequency, Monetary)

Primero hay que seleccionar la información de interes:
- Store
- Date
- Weekly_Sales

Después hay que quitar nulos y agrupar los datos por Date y Store  y hacer la suma de Weekly_Sales.

In [0]:
RFM0 = (featuresSalesStores
       .select('Store', 'Date', 'Weekly_Sales')
       .dropna('any')
       .groupby(['Store','Date'])
       .agg(sf.sum('Weekly_Sales').alias('Weekly_Sales'))
       .withColumn('Weekly_Sales', sf.round(col('Weekly_Sales'), 2))
       )

RFM0.display()

Store Date Weekly_Sales 31 2011-06-24 1375962.46 28 2011-08-26 1166479.51 28 2011-05-13 1253316.3 28 2011-09-23 1109105.92 26 2011-09-02 1040143.14 27 2012-02-10 1651605.35 27 2011-04-15 1727175.61 1 2010-04-02 1594968.28 1 2010-03-05 1554806.68 1 2011-12-16 1881176.67 6 2011-10-07 1481739.2 16 2010-03-05 444181.85 3 2012-10-26 405432.7 40 2011-12-16 1198670.19 15 2012-09-07 587259.82 43 2011-09-23 586108.13 35 2010-03-12 1150344.39 4 2012-07-27 2048613.65 4 2011-10-21 2207742.13 4 2010-04-30 1817273.28 8 2011-11-18 897032.19 45 2012-06-29 773367.71 38 2010-06-11 354828.65 25 2010-11-26 1115240.61 25 2012-09-28 683300.84 24 2012-08-31 1344558.92 32 2011-07-15 1144254.26 33 2011-05-13 279466.87 33 2011-02-25 242901.21 33 2010-03-19 282235.73 33 2010-07-02 267495.76 42 2010-07-23 491115.86 2 2012-03-30 1857480.84 2 2012-07-27 1757923.88 2 2010-03-26 1750197.81 18 2011-02-18 1095058.57 18 2010-08-27 1214302.76 18 2011-06-17 1045859.39 31 2012-08-24 1379783.21 31 2012-07-27 1303732.36 28 2010-06-18 1267619.06 28 2011-05-27 1160043.98 27 2012-07-06 2062224.92 12 2011-07-29 890547.07 12 2010-12-31 891736.91 12 2010-09-03 1075758.55 22 2011-07-29 944958.69 13 2012-10-05 2041918.74 6 2010-11-19 1518841.45 3 2010-04-30 361248.39 20 2011-10-14 2010107.68 40 2012-09-07 1088248.4 5 2010-07-02 305993.27 37 2012-10-19 551969.1 23 2011-11-25 2057059.53 7 2012-01-27 541037.98 10 2010-10-15 1703850.25 45 2011-10-14 720946.99 24 2012-08-03 1440374.13 32 2011-07-01 1199845.29 14 2011-06-03 2079899.47 42 2012-02-10 674919.45 34 2011-07-29 851461.9 27 2011-04-22 1921655.48 44 2012-07-20 337819.16 12 2010-11-05 929690.71 13 2010-04-30 1785823.37 41 2012-06-15 1384584.59 15 2010-10-01 566945.95 37 2010-12-24 576809.92 8 2010-11-19 885608.04 23 2012-05-18 1332952.47 10 2010-10-08 1772192.42 25 2012-03-30 684348.92 24 2010-05-07 1409705.03 29 2011-10-14 475776.45 29 2010-03-05 535087.91 29 2010-12-03 642678.53 32 2012-03-30 1108686.87 11 2011-09-23 1235775.15 11 2012-05-04 1370251.22 33 2011-06-10 258427.39 14 2011-02-25 1875708.88 14 2012-10-05 1687592.16 14 2012-04-13 1912909.69 30 2011-04-01 437926.79 30 2010-07-30 437893.76 36 2011-02-18 427175.03 27 2010-07-23 1808250.71 44 2012-03-23 296947.06 22 2010-10-29 1025766.27 22 2011-10-14 944337.32 6 2012-07-06 1876359.39 6 2010-05-07 1619920.04 3 2012-05-11 431985.36 20 2012-04-06 2565259.92 40 2010-11-26 1166142.85 19 2010-08-06 1492060.89 41 2012-02-17 1330451.46 41 2010-03-19 1006597.69 15 2010-07-16 633203.69 15 2012-02-03 523831.64 9 2010-02-05 549505.55 35 2011-02-11 859258.17 4 2011-04-22 2220600.76 8 2012-02-24 928537.54 45 2012-07-20 737613.65 25 2011-03-04 640912.18 24 2011-08-26 1402372.09 29 2010-04-16 509100.84 32 2010-11-05 1063056.21 14 2012-10-12 1639585.61 18 2010-04-02 1254107.84 31 2012-09-28 1279080.58 28 2010-06-25 1231025.07 26 2010-07-09 1122356.53 27 2012-09-07 1840955.23 27 2012-10-05 1591816.88 12 2012-04-06 1128765.71 1 2011-02-18 1686842.78 13 2010-03-12 1840686.94 13 2010-02-19 1970274.64 6 2011-04-29 1479249.9 20 2012-02-03 2203523.2 40 2012-06-15 977950.28 5 2010-08-13 286477.35 5 2010-07-16 280701.7 41 2011-10-21 1214944.29 15 2010-11-12 619639.74 37 2012-10-12 521810.75 9 2010-07-23 462676.5 4 2010-08-13 2007050.75 8 2012-05-11 920128.89 8 2012-09-14 883569.38 8 2012-02-17 1096232.89 39 2010-10-22 1261109.01 39 2010-11-12 1291398.71 7 2012-08-24 609099.37 45 2010-05-07 812190.76 38 2011-09-09 397771.68 38 2011-08-26 365672.55 24 2012-05-18 1298809.8 24 2012-01-20 1151993.85 24 2011-02-18 1360150.12 21 2010-05-21 753361.72 11 2010-07-23 1244390.03 14 2012-03-09 2020839.31 14 2010-07-16 2130287.27 14 2011-05-06 2095599.93 42 2010-08-20 500945.63 30 2011-07-01 428782.38 36 2011-02-25 380279.44 31 2010-05-28 1309437.17 31 2012-10-05 1363365.05 28 2012-03-02 1451740.57 28 2012-03-30 1209524.11 44 2010-03-12 272190.83 12 2011-05-13 977033.5 22 2011-11-11 1051740.29 19 2010-06-04 1519013.49 41 2011-09-09 1280958.97 43 2010-04-23 638957.35 43 2011-08-12 595626.56 4 2010-09-03 

### Recency

Primero se obtiene la última fecha a analizar, en este caso será la útlima reportada.

In [0]:
max_date = RFM0.agg(sf.max('Date'))
max_date.display()

maxDate = max_date.groupBy('max(Date)').max().collect()[0] # Para tomar el valor del DataFrame

max(Date) 2012-10-26

In [0]:
max_date.printSchema()

root
-- max(Date): date (nullable = true)

Luego se revisa la última fecha registrada por cada tienda. En este caso se observa que todas las tiendas han comprado en la misma fecha que la fecha máxima global.

In [0]:
storesMaxDate = RFM0.groupby(['Store']).agg(sf.max('Date'))
display(storesMaxDate)

Store max(Date) 31 2012-10-26 34 2012-10-26 28 2012-10-26 26 2012-10-26 27 2012-10-26 44 2012-10-26 12 2012-10-26 22 2012-10-26 1 2012-10-26 13 2012-10-26 6 2012-10-26 16 2012-10-26 3 2012-10-26 40 2012-10-26 20 2012-10-26 5 2012-10-26 19 2012-10-26 41 2012-10-26 15 2012-10-26 43 2012-10-26 37 2012-10-26 9 2012-10-26 17 2012-10-26 35 2012-10-26 4 2012-10-26 8 2012-10-26 23 2012-10-26 39 2012-10-26 7 2012-10-26 10 2012-10-26 45 2012-10-26 38 2012-10-26 25 2012-10-26 24 2012-10-26 29 2012-10-26 21 2012-10-26 32 2012-10-26 33 2012-10-26 11 2012-10-26 14 2012-10-26 42 2012-10-26 2 2012-10-26 30 2012-10-26 18 2012-10-26 36 2012-10-26

In [0]:
storesMaxDate.printSchema()

root
-- Store: integer (nullable = true)
-- max(Date): date (nullable = true)

Obteniendo la diferencia entre la máxima fecha global y la máxima fecha de cada tienda:

In [0]:
r = (RFM0
     .join(storesMaxDate, on = 'Store')
     .withColumn('Recency', datediff(to_date(lit(maxDate[0])), col('max(Date)')))
    )

display(r)
#recency.count()

Store Date Weekly_Sales max(Date) Recency 31 2011-06-24 1375962.46 2012-10-26 0 31 2012-08-24 1379783.21 2012-10-26 0 31 2012-07-27 1303732.36 2012-10-26 0 31 2012-09-28 1279080.58 2012-10-26 0 31 2010-05-28 1309437.17 2012-10-26 0 31 2012-10-05 1363365.05 2012-10-26 0 31 2011-12-23 2026176.14 2012-10-26 0 31 2012-08-10 1386472.59 2012-10-26 0 31 2010-04-16 1367448.28 2012-10-26 0 31 2011-09-23 1347607.74 2012-10-26 0 31 2010-09-03 1302047.48 2012-10-26 0 31 2011-07-15 1415473.91 2012-10-26 0 31 2012-05-04 1391257.28 2012-10-26 0 31 2010-08-20 1376571.21 2012-10-26 0 31 2012-02-10 1527688.58 2012-10-26 0 31 2011-08-19 1487542.53 2012-10-26 0 31 2012-05-25 1397094.26 2012-10-26 0 31 2010-06-18 1368090.08 2012-10-26 0 31 2012-08-31 1373651.49 2012-10-26 0 31 2010-04-09 1395710.09 2012-10-26 0 31 2011-01-21 1344167.13 2012-10-26 0 31 2011-03-25 1357036.1 2012-10-26 0 31 2010-06-04 1427624.5 2012-10-26 0 31 2011-05-20 1387365.19 2012-10-26 0 31 2010-05-07 1381796.8 2012-10-26 0 31 2012-02-17 1570813.52 2012-10-26 0 31 2010-04-23 1280465.8 2012-10-26 0 31 2010-12-03 1338716.37 2012-10-26 0 31 2011-11-11 1444783.64 2012-10-26 0 31 2012-01-13 1377485.12 2012-10-26 0 31 2011-11-25 1934099.65 2012-10-26 0 31 2011-07-29 1308122.15 2012-10-26 0 31 2012-06-22 1394065.76 2012-10-26 0 31 2010-07-30 1231993.14 2012-10-26 0 31 2010-04-30 1243200.24 2012-10-26 0 31 2010-10-08 1300131.68 2012-10-26 0 31 2011-01-07 1311986.87 2012-10-26 0 31 2010-09-24 1230250.25 2012-10-26 0 31 2010-03-05 1384870.51 2012-10-26 0 31 2012-07-20 1392395.2 2012-10-26 0 31 2011-11-04 1426405.46 2012-10-26 0 31 2011-08-12 1443311.49 2012-10-26 0 31 2012-07-06 1458059.42 2012-10-26 0 31 2010-02-19 1473386.75 2012-10-26 0 31 2012-06-29 1349202.25 2012-10-26 0 31 2010-09-17 1294105.01 2012-10-26 0 31 2011-01-28 1254955.68 2012-10-26 0 31 2012-09-21 1373064.87 2012-10-26 0 31 2012-04-20 1407036.59 2012-10-26 0 31 2010-03-19 1332261.01 2012-10-26 0 31 2011-04-22 1461393.91 2012-10-26 0 31 2012-07-13 1374301.34 2012-10-26 0 31 2011-02-25 1365678.22 2012-10-26 0 31 2011-03-04 1412721.18 2012-10-26 0 31 2010-05-21 1337405.6 2012-10-26 0 31 2012-06-01 1404516.29 2012-10-26 0 31 2011-05-27 1342273.64 2012-10-26 0 31 2010-08-27 1326621.98 2012-10-26 0 31 2011-07-08 1423144.47 2012-10-26 0 31 2012-01-27 1259941.48 2012-10-26 0 31 2012-01-06 1401232.52 2012-10-26 0 31 2010-05-14 1287288.59 2012-10-26 0 31 2011-12-09 1556798.94 2012-10-26 0 31 2011-10-28 1346862.72 2012-10-26 0 31 2010-07-02 1311704.92 2012-10-26 0 31 2010-02-12 1543947.23 2012-10-26 0 31 2010-12-24 2068942.97 2012-10-26 0 31 2012-02-03 1391479.91 2012-10-26 0 31 2010-03-12 1366193.35 2012-10-26 0 31 2012-02-24 1392543.37 2012-10-26 0 31 2012-09-07 1358111.62 2012-10-26 0 31 2010-12-10 1475685.1 2012-10-26 0 31 2011-07-01 1371889.27 2012-10-26 0 31 2010-10-29 1257921.28 2012-10-26 0 31 2011-12-30 1355405.95 2012-10-26 0 31 2010-09-10 1308179.02 2012-10-26 0 31 2011-09-16 1363460.86 2012-10-26 0 31 2011-03-11 1405572.93 2012-10-26 0 31 2010-11-12 1352780.76 2012-10-26 0 31 2010-03-26 1229635.7 2012-10-26 0 31 2012-09-14 1327705.44 2012-10-26 0 31 2012-10-26 1340232.55 2012-10-26 0 31 2011-04-01 1344890.73 2012-10-26 0 31 2012-03-23 1365824.97 2012-10-26 0 31 2010-08-13 1327719.34 2012-10-26 0 31 2011-05-13 1408118.22 2012-10-26 0 31 2011-08-05 1428993.33 2012-10-26 0 31 2012-03-09 1439034.86 2012-10-26 0 31 2012-08-03 1390174.63 2012-10-26 0 31 2012-04-27 1311352.25 2012-10-26 0 31 2012-10-19 1378730.45 2012-10-26 0 31 2011-08-26 1417267.07 2012-10-26 0 31 2010-10-01 1213981.64 2012-10-26 0 31 2010-12-17 1714667.0 2012-10-26 0 31 2012-03-16 1481728.13 2012-10-26 0 31 2012-08-17 1410181.7 2012-10-26 0 31 2012-04-13 1407842.91 2012-10-26 0 31 2010-08-06 1352401.08 2012-10-26 0 31 2010-07-23 1276031.84 2012-10-26 0 31 2012-10-12 1401113.42 2012-10-26 0 31 2010-11-05 1324455.72 2012-10-26 0 31 2011-06-03 1430348.1 2012-10-26 0 31 2012-05-11 1392938.06 2012-10-26 0 31 2012-03-02 1427881.22 2012-10-26 0 31 2011-10-14 137

### Frequency

Se obtiene la frecuencia de compras que hubo en el periodo dado. En este caso se observa como todas las tiendas tuvieron ventas todos los días.

In [0]:
f = (RFM0
    .groupby(['Store'])
    .count()
    .withColumn('Frequency',col('count'))
    )

f.display()

Store count Frequency 31 143 143 34 143 143 28 143 143 26 143 143 27 143 143 44 143 143 12 143 143 22 143 143 1 143 143 13 143 143 6 143 143 16 143 143 3 143 143 40 143 143 20 143 143 5 143 143 19 143 143 41 143 143 15 143 143 43 143 143 37 143 143 9 143 143 17 143 143 35 143 143 4 143 143 8 143 143 23 143 143 39 143 143 7 143 143 10 143 143 45 143 143 38 143 143 25 143 143 24 143 143 29 143 143 21 143 143 32 143 143 33 143 143 11 143 143 14 143 143 42 143 143 2 143 143 30 143 143 18 143 143 36 143 143

### Monetary

In [0]:
m = (RFM0
    .groupby('Store')
    .agg(sf.sum('Weekly_Sales').alias('Monetary'))
    .withColumn('Monetary', sf.round(col('Monetary'), 2))
    )

m.display()

Store Monetary 31 1.996139055E8 34 1.38249763E8 28 1.8926368058E8 26 1.4341639379E8 27 2.5385591688E8 44 4.329308784E7 12 1.4428723015E8 22 1.4707564857E8 1 2.2240280885E8 13 2.865177038E8 6 2.2375613064E8 16 7.42524254E7 3 5.758673507E7 40 1.3787030979E8 20 3.0139779246E8 5 4.54756889E7 19 2.066348621E8 41 1.8134193489E8 15 8.913368392E7 43 9.056543541E7 37 7.420274032E7 9 7.778921899E7 17 1.2778213883E8 35 1.3152067208E8 4 2.9954395338E8 8 1.2995118113E8 23 1.9875061785E8 39 2.0744554247E8 7 8.159827514E7 10 2.7161771389E8 45 1.1239534142E8 38 5.515962642E7 25 1.0106117917E8 24 1.9401602128E8 29 7.714155431E7 21 1.0811787892E8 32 1.6681924616E8 33 3.716022196E7 11 1.939627868E8 14 2.8899991134E8 42 7.956575243E7 2 2.7538244098E8 30 6.271688512E7 18 1.5511473421E8 36 5.341221497E7

### RFM

Juntando los DataFrames r, f, m con las tiendas.

In [0]:
RFM1 = (r.select('Store', 'Recency').distinct()
       .join(f.select('Store', 'Frequency'), on = 'Store')
       .join(m, on = 'Store')
       )

RFM1.display()

Store Recency Frequency Monetary 31 0 143 1.996139055E8 34 0 143 1.38249763E8 28 0 143 1.8926368058E8 26 0 143 1.4341639379E8 27 0 143 2.5385591688E8 44 0 143 4.329308784E7 12 0 143 1.4428723015E8 22 0 143 1.4707564857E8 1 0 143 2.2240280885E8 13 0 143 2.865177038E8 6 0 143 2.2375613064E8 16 0 143 7.42524254E7 3 0 143 5.758673507E7 20 0 143 3.0139779246E8 40 0 143 1.3787030979E8 5 0 143 4.54756889E7 19 0 143 2.066348621E8 41 0 143 1.8134193489E8 15 0 143 8.913368392E7 43 0 143 9.056543541E7 37 0 143 7.420274032E7 9 0 143 7.778921899E7 17 0 143 1.2778213883E8 35 0 143 1.3152067208E8 4 0 143 2.9954395338E8 8 0 143 1.2995118113E8 23 0 143 1.9875061785E8 39 0 143 2.0744554247E8 7 0 143 8.159827514E7 10 0 143 2.7161771389E8 45 0 143 1.1239534142E8 38 0 143 5.515962642E7 25 0 143 1.0106117917E8 24 0 143 1.9401602128E8 29 0 143 7.714155431E7 21 0 143 1.0811787892E8 32 0 143 1.6681924616E8 11 0 143 1.939627868E8 33 0 143 3.716022196E7 14 0 143 2.8899991134E8 42 0 143 7.956575243E7 2 0 143 2.7538244098E8 30 0 143 6.271688512E7 18 0 143 1.5511473421E8 36 0 143 5.341221497E7

## Saving data

In [0]:
#%sh
#pwd
#ls 
display(dbutils.fs.ls("FileStore/tables"))

path name size dbfs:/FileStore/tables/Features_data_set-1.csv Features_data_set-1.csv 600478 dbfs:/FileStore/tables/Features_data_set.csv Features_data_set.csv 600478 dbfs:/FileStore/tables/sales_data_set-1.csv sales_data_set-1.csv 13264115 dbfs:/FileStore/tables/sales_data_set.csv sales_data_set.csv 13264115 dbfs:/FileStore/tables/stores_data_set-1.csv stores_data_set-1.csv 577 dbfs:/FileStore/tables/stores_data_set.csv stores_data_set.csv 577

In [0]:
workingDir = "FileStore/tables"
targetPath = f"{workingDir}/RFM"

RFM1.write.mode("OVERWRITE").parquet(targetPath)

In [0]:
display(dbutils.fs.ls(targetPath))

path name size dbfs:/FileStore/tables/RFM/_SUCCESS _SUCCESS 0 dbfs:/FileStore/tables/RFM/_committed_7903083045350027116 _committed_7903083045350027116 4002 dbfs:/FileStore/tables/RFM/_started_7903083045350027116 _started_7903083045350027116 0 dbfs:/FileStore/tables/RFM/part-00000-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43074-1-c000.snappy.parquet part-00000-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43074-1-c000.snappy.parquet 572 dbfs:/FileStore/tables/RFM/part-00002-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43053-1-c000.snappy.parquet part-00002-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43053-1-c000.snappy.parquet 1184 dbfs:/FileStore/tables/RFM/part-00011-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43067-1-c000.snappy.parquet part-00011-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43067-1-c000.snappy.parquet 1183 dbfs:/FileStore/tables/RFM/part-00014-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43045-1-c000.snappy.parquet part-00014-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43045-1-c000.snappy.parquet 1183 dbfs:/FileStore/tables/RFM/part-00019-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43042-1-c000.snappy.parquet part-00019-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43042-1-c000.snappy.parquet 1260 dbfs:/FileStore/tables/RFM/part-00021-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43050-1-c000.snappy.parquet part-00021-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43050-1-c000.snappy.parquet 1184 dbfs:/FileStore/tables/RFM/part-00024-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43046-1-c000.snappy.parquet part-00024-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43046-1-c000.snappy.parquet 1184 dbfs:/FileStore/tables/RFM/part-00030-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43051-1-c000.snappy.parquet part-00030-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43051-1-c000.snappy.parquet 1184 dbfs:/FileStore/tables/RFM/part-00043-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43072-1-c000.snappy.parquet part-00043-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43072-1-c000.snappy.parquet 1184 dbfs:/FileStore/tables/RFM/part-00048-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43056-1-c000.snappy.parquet part-00048-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43056-1-c000.snappy.parquet 1184 dbfs:/FileStore/tables/RFM/part-00049-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43038-1-c000.snappy.parquet part-00049-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43038-1-c000.snappy.parquet 1260 dbfs:/FileStore/tables/RFM/part-00051-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43055-1-c000.snappy.parquet part-00051-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43055-1-c000.snappy.parquet 1184 dbfs:/FileStore/tables/RFM/part-00053-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43041-1-c000.snappy.parquet part-00053-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43041-1-c000.snappy.parquet 1260 dbfs:/FileStore/tables/RFM/part-00066-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43057-1-c000.snappy.parquet part-00066-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43057-1-c000.snappy.parquet 1184 dbfs:/FileStore/tables/RFM/part-00069-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43061-1-c000.snappy.parquet part-00069-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43061-1-c000.snappy.parquet 1184 dbfs:/FileStore/tables/RFM/part-00075-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43068-1-c000.snappy.parquet part-00075-tid-7903083045350027116-f21a6edd-d7a5-4227-b95c-ab5fd242fd80-43068-1-c000.snappy.parquet 1184 dbfs:/FileStore/tables/RFM/part-00077-tid-7903